<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="wolfemd" />

<meta name="date" content="2022-04-20" />

<title>Prelminary field trial analysis</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GSexample2022</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="download.html">Download</a>
</li>
<li>
  <a href="clean_phenos.html">Clean phenos</a>
</li>
<li>
  <a href="prep_genos.html">Prep genos</a>
</li>
<li>
  <a href="get_blups.html">Get BLUPs</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Prelminary field trial analysis</h1>
<h4 class="author">wolfemd</h4>
<h4 class="date">2022-04-20</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-04-20
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>GSexample2022/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220320code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20220320)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220320code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220320)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomwolfemdGSexample2022tree9c0e19e38ee5c830efc736987c5fd8e9e6450616targetblank9c0e19ea"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/wolfemd/GSexample2022/tree/9c0e19e38ee5c830efc736987c5fd8e9e6450616" target="_blank">9c0e19e</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomwolfemdGSexample2022tree9c0e19e38ee5c830efc736987c5fd8e9e6450616targetblank9c0e19ea" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/wolfemd/GSexample2022/tree/9c0e19e38ee5c830efc736987c5fd8e9e6450616" target="_blank">9c0e19e</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  .DS_Store
    Untracked:  Icon
    Untracked:  analysis/.DS_Store
    Untracked:  analysis/Icon
    Untracked:  analysis/images/Icon
    Untracked:  analysis/images/paste-2F34C15B.png
    Untracked:  analysis/images/paste-AD7AEEAC.png
    Untracked:  analysis/images/paste-CB5D91AC.png
    Untracked:  analysis/troubleshooting.Rmd
    Untracked:  code/Icon
    Untracked:  data/.DS_Store
    Untracked:  data/BreedBaseGenotypesDownload.positions
    Untracked:  data/BreedBaseGenotypesDownload.vcf
    Untracked:  data/BreedBaseGenotypes_subset.hap.gz
    Untracked:  data/BreedBaseGenotypes_subset.positions
    Untracked:  data/BreedBaseGenotypes_subset.samples
    Untracked:  data/Icon
    Untracked:  data/metadata_cynthia.csv
    Untracked:  data/phenotype_cynthia.csv
    Untracked:  data/subset_unique_names_for_vcf.txt
    Untracked:  data/unique_names_for_vcf.txt
    Untracked:  data/vcf_colnames.txt
    Untracked:  out.log
    Untracked:  output/.DS_Store
    Untracked:  output/Icon
    Untracked:  output/blups.rds
    Untracked:  output/models_compared.Rdata

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/get_blups.Rmd</code>) and HTML (<code>docs/get_blups.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/9c0e19e38ee5c830efc736987c5fd8e9e6450616/analysis/get_blups.Rmd" target="_blank">9c0e19e</a>
</td>
<td>
wolfemd
</td>
<td>
2022-04-20
</td>
<td>
Add the ‘preliminary trial analysis’ or ‘get blups’ steps.
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Follow this worked example <em>and</em> reference the <a href="https://wolfemd.github.io/GenomicSelectionManual/preliminary-field-trial-analysis.html">“Genomic Selection Manual” - Chapter 9</a>.</p>
<div id="load-packages" class="section level1">
<h1>Load packages</h1>
<pre class="r"><code>suppressMessages(library(tidyverse)); 
suppressMessages(library(genomicMateSelectR)); 
suppressMessages(library(sommer));</code></pre>
</div>
<div id="set-up-the-training-data" class="section level1">
<h1>Set-up the training data</h1>
<pre class="r"><code>phenos&lt;-readRDS(here::here(&quot;output&quot;,&quot;phenotypes_cleaned.rds&quot;))</code></pre>
<pre class="r"><code>phenos %&gt;% 
     count(CompleteBlocks,IncompleteBlocks,locationName) %&gt;% 
     spread(locationName,n)</code></pre>
<pre><code># A tibble: 3 × 4
  CompleteBlocks IncompleteBlocks Ibadan Ubiaja
  &lt;lgl&gt;          &lt;lgl&gt;             &lt;int&gt;  &lt;int&gt;
1 FALSE          TRUE                777    792
2 TRUE           FALSE                NA    125
3 TRUE           TRUE                407    414</code></pre>
<p>Convert the data to “long format” . Remove missing values. “Nest” the data by Trait.</p>
<pre class="r"><code>traits&lt;-c(&quot;DM&quot;,&quot;MCMDS&quot;,&quot;logFYLD&quot;,&quot;logDYLD&quot;)
phenos&lt;-phenos %&gt;% 
     # Convert the data to &quot;long format&quot; 
     pivot_longer(cols = all_of(traits), 
                  names_to = &quot;Trait&quot;, values_to = &quot;Value&quot;) %&gt;%
     # Remove missing values
     filter(!is.na(Value)) %&gt;%
     # Nest the MultiEnvironmentTrial data by trait
     nest(METdata=c(-Trait))</code></pre>
<pre class="r"><code>phenos %&gt;% 
     mutate(N_plots=map_dbl(METdata,nrow))</code></pre>
<pre><code># A tibble: 4 × 3
  Trait   METdata               N_plots
  &lt;chr&gt;   &lt;list&gt;                  &lt;dbl&gt;
1 DM      &lt;tibble [2,211 × 35]&gt;    2211
2 MCMDS   &lt;tibble [1,436 × 35]&gt;    1436
3 logFYLD &lt;tibble [1,590 × 35]&gt;    1590
4 logDYLD &lt;tibble [1,443 × 35]&gt;    1443</code></pre>
<pre class="r"><code>phenos$METdata[[1]] %&gt;% head</code></pre>
<pre><code># A tibble: 6 × 35
  studyYear programName locationName studyName  studyDesign plotWidth plotLength
      &lt;int&gt; &lt;chr&gt;       &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;           &lt;int&gt;      &lt;dbl&gt;
1      2017 IITA        Ubiaja       17.GS.Gen… Augmented           1          5
2      2017 IITA        Ubiaja       17.GS.Gen… Augmented           1          5
3      2017 IITA        Ubiaja       17.GS.Gen… Augmented           1          5
4      2017 IITA        Ubiaja       17.GS.Gen… Augmented           1          5
5      2017 IITA        Ubiaja       17.GS.Gen… Augmented           1          5
6      2017 IITA        Ubiaja       17.GS.Gen… Augmented           1          5
# … with 28 more variables: fieldSize &lt;dbl&gt;, plantingDate &lt;chr&gt;,
#   harvestDate &lt;chr&gt;, germplasmName &lt;chr&gt;, observationUnitDbId &lt;int&gt;,
#   replicate &lt;int&gt;, blockNumber &lt;int&gt;, plotNumber &lt;int&gt;, rowNumber &lt;int&gt;,
#   colNumber &lt;int&gt;, entryType &lt;chr&gt;, trialType &lt;chr&gt;, numberBlocks &lt;int&gt;,
#   numberReps &lt;int&gt;, observationUnitName &lt;chr&gt;, CompleteBlocks &lt;lgl&gt;,
#   IncompleteBlocks &lt;lgl&gt;, yearInLoc &lt;chr&gt;, trialInLocYr &lt;chr&gt;,
#   repInTrial &lt;chr&gt;, blockInRep &lt;chr&gt;, NOHAV &lt;dbl&gt;, plotArea &lt;dbl&gt;, …</code></pre>
</div>
<div id="comparison-of-models" class="section level1">
<h1>Comparison of models</h1>
<pre class="r"><code># grab just one chunk of the data to &quot;experiment&quot; on
METdata&lt;-phenos$METdata[[1]] %&gt;% 
     # sommer doesn&#39;t recognize logical T/F variables
     mutate(CompleteBlocks=as.character(CompleteBlocks), 
            IncompleteBlocks=as.character(IncompleteBlocks))</code></pre>
<p>How to run a single, simple model, <em>not</em> using future.</p>
<p>This next step takes at least 25 minutes to run on my laptop.</p>
<pre class="r"><code>library(sommer)
library(future);
plan(multisession)

# the model I&#39;ve applied to NextGen predictions the last several years
conv_model %&lt;-% { mmer(Value~yearInLoc,
                       random=~germplasmName + 
                            vs(at(CompleteBlocks,&quot;TRUE&quot;),repInTrial) + 
                            vs(at(IncompleteBlocks,&quot;TRUE&quot;),blockInRep),
                       data=METdata, 
                       verbose = FALSE) }
# add heterogeneous-error variance for each trial (studyName)
conv_het_error_model %&lt;-% { mmer(Value~yearInLoc,
                                 random=~germplasmName + 
                                      vs(at(CompleteBlocks,&quot;TRUE&quot;),repInTrial) + 
                                      vs(at(IncompleteBlocks,&quot;TRUE&quot;),blockInRep),
                                 rcov=~vs(ds(studyName),units),
                                 data=METdata, 
                                 verbose = FALSE) }
# simplify: no het. error, no at() variance structure
simple_model %&lt;-% { mmer(Value~yearInLoc,
                         random=~germplasmName + repInTrial + blockInRep,
                         data=METdata, 
                         verbose = FALSE) }
# no at() variance structure, include het. error variance by trial
simple_het_error_model %&lt;-% { mmer(Value~yearInLoc,
                                   random=~germplasmName + repInTrial + blockInRep,
                                   rcov=~vs(ds(studyName),units),
                                   data=METdata, 
                                   verbose = FALSE) }</code></pre>
<pre class="r"><code># Save output of above so the above models don&#39;t need to be re-run
## if and when I reformat or otherwise re-knit this doc to HTML
save(conv_model,conv_het_error_model,simple_model,simple_het_error_model,
     file=here::here(&quot;output&quot;,&quot;models_compared.Rdata&quot;))</code></pre>
<p>Load the fitted model outputs.</p>
<pre class="r"><code># load the output from the models from where I stored it
load(here::here(&quot;output&quot;,&quot;models_compared.Rdata&quot;))</code></pre>
<p>Which model fits best? How to compare?</p>
<p>sommer provides <code>anova()</code> for pairwise comparisons of models using a likelihood ratio test and by AIC, BIC.</p>
<pre class="r"><code>anova(conv_model,conv_het_error_model)</code></pre>
<pre><code>Likelihood ratio test for mixed models
==============================================================
     Df       AIC      BIC     loLik     Chisq ChiDf                  PrChisq
mod2 21  87.83861 104.9422  -40.9193                                         
mod1 13 567.09329 584.1969 -280.5466 479.25469     8 1.98186311070714e-98 ***
==============================================================
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<pre><code>     Df       AIC      BIC     loLik     Chisq ChiDf                  PrChisq
mod2 21  87.83861 104.9422  -40.9193                                         
mod1 13 567.09329 584.1969 -280.5466 479.25469     8 1.98186311070714e-98 ***</code></pre>
<p>Here’s a very not-pretty not-high-throughput way to set-up a table comparing the AIC for all 4 models.</p>
<pre class="r"><code>tibble(Model=c(&quot;conv_model&quot;,&quot;conv_het_error_model&quot;,
               &quot;simple_model&quot;,&quot;simple_het_error_model&quot;),
       AIC=c(conv_model$AIC,conv_het_error_model$AIC,
             simple_model$AIC,simple_het_error_model$AIC)) %&gt;% 
     arrange(AIC)</code></pre>
<pre><code># A tibble: 4 × 2
  Model                    AIC
  &lt;chr&gt;                  &lt;dbl&gt;
1 simple_het_error_model  81.1
2 conv_het_error_model    87.8
3 simple_model           560. 
4 conv_model             567. </code></pre>
<p>The “simple” heterogeneous error model fits the data best, I’ll go ahead and use it for this demonstration.</p>
<div id="activity" class="section level2">
<h2>ACTIVITY</h2>
<p>Run the simple model, examine the results.</p>
<p>Option 1) Use <code>sommer</code>. Takes ~9 mins on my laptop.</p>
<pre class="r"><code>simple_model&lt;-mmer(Value~yearInLoc,
                   random=~germplasmName + repInTrial + blockInRep,
                   data=METdata, 
                   verbose = TRUE)</code></pre>
<p>Option 2) Use package <code>lme4</code>’s <code>lmer()</code> function… takes seconds. Same result.</p>
<p>Again, note that <code>sommer</code> advantage is flexible modeling.</p>
<p>Plus, <code>sommer</code> can do genomic prediction, <code>lmer</code> can’t do that.</p>
<pre class="r"><code>library(lme4)
simple_model_lmer&lt;-lmer(Value~yearInLoc + (1|germplasmName) + (1|repInTrial) + (1|blockInRep),
                        data=METdata)

summary(simple_model_lmer)</code></pre>
<pre><code>Linear mixed model fit by REML [&#39;lmerMod&#39;]
Formula: Value ~ yearInLoc + (1 | germplasmName) + (1 | repInTrial) +  
    (1 | blockInRep)
   Data: METdata

REML criterion at convergence: 13150.4

Scaled residuals: 
    Min      1Q  Median      3Q     Max 
-3.8460 -0.4014  0.0359  0.4478  3.3480 

Random effects:
 Groups        Name        Variance Std.Dev.
 germplasmName (Intercept) 16.365   4.045   
 blockInRep    (Intercept)  1.059   1.029   
 repInTrial    (Intercept)  5.855   2.420   
 Residual                   9.696   3.114   
Number of obs: 2211, groups:  
germplasmName, 1531; blockInRep, 137; repInTrial, 16

Fixed effects:
                          Estimate Std. Error t value
(Intercept)                34.7765     0.9717  35.789
yearInLocIITA_Ubiaja_2017  -7.8778     2.6315  -2.994
yearInLocIITA_Ubiaja_2019  -1.9611     1.3178  -1.488

Correlation of Fixed Effects:
               (Intr) yILIITA_U_2017
yILIITA_U_2017 -0.368               
yILIITA_U_2019 -0.707  0.261        </code></pre>
</div>
</div>
<div id="analyze-all-traits" class="section level1">
<h1>Analyze all traits</h1>
<p>Next we will fit the “best” model from the previous step, but to all traits.</p>
<p>We have four data chunks set-up in the tibble <code>phenos</code></p>
<pre class="r"><code>phenos</code></pre>
<pre><code># A tibble: 4 × 2
  Trait   METdata              
  &lt;chr&gt;   &lt;list&gt;               
1 DM      &lt;tibble [2,211 × 35]&gt;
2 MCMDS   &lt;tibble [1,436 × 35]&gt;
3 logFYLD &lt;tibble [1,590 × 35]&gt;
4 logDYLD &lt;tibble [1,443 × 35]&gt;</code></pre>
<p>The manual, starting <a href="https://wolfemd.github.io/GenomicSelectionManual/preliminary-field-trial-analysis.html#test-code-for-function">here</a> walks you through the steps creating and debugging a function to analyze all traits. That function is in an R script called <code>get_blups.R</code> in the <code>code/</code> folder. We will use it below.</p>
<pre class="r"><code># Source the R script
source(here::here(&quot;code&quot;,&quot;get_blups.R&quot;))
# Set-up inputs for the function
phenos %&lt;&gt;% 
     mutate(fixedFormula=ifelse(Trait %in% c(&quot;logDYLD&quot;,&quot;logFYLD&quot;,&quot;logRTNO&quot;,&quot;logTOPYLD&quot;),
                                &quot;Value ~ yearInLoc + PropNOHAV&quot;,
                                &quot;Value ~ yearInLoc&quot;),
            randFormula=&quot;~germplasmName + repInTrial + blockInRep&quot;,
            rcovFormula=&quot;~vs(ds(studyName),units)&quot;)</code></pre>
<p>This next step took <strong><em>45 minutes</em></strong> on my laptop.</p>
<pre class="r"><code>starttime&lt;-proc.time()[3]
require(furrr); plan(multisession, workers = 4)
blups&lt;-phenos %&gt;%
     mutate(modelOut=future_pmap(.,get_blups)) %&gt;% 
     unnest(modelOut)
stoptime&lt;-proc.time()[3]; elapsed&lt;-stoptime-starttime; elapsed/60
#  elapsed 
# 44.86077
# Save
saveRDS(blups,file = here::here(&quot;output&quot;,&quot;blups.rds&quot;))</code></pre>
<pre class="r"><code>blups&lt;-readRDS(here::here(&quot;output&quot;,&quot;blups.rds&quot;))
# Print the output, add a few summaries, like H2
blups %&gt;% 
     dplyr::select(Trait,H2,meanHetVarE,meanNreps,outliers) %&gt;% 
     mutate(H2=round(H2,3),
            meanNreps=round(meanNreps,3),
            outliers=map_dbl(outliers,length))</code></pre>
<pre><code># A tibble: 4 × 5
  Trait      H2 meanHetVarE meanNreps outliers
  &lt;chr&gt;   &lt;dbl&gt;       &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 DM      0.426       9.80       1.44        8
2 MCMDS   0           0.205      1.52       33
3 logFYLD 0.3         0.129      1.72        9
4 logDYLD 0.347       0.112      1.78        4</code></pre>
<p>That’s the end of the core steps.</p>
<p>The rest of the manual chapter includes some exploration of the results <em>and</em> explanation of BLUPs, de-regressed BLUPs and weighted analysis.</p>
<p>I will try to cover that with some slides to give you intuition.</p>
</div>
<div id="study-material" class="section level1">
<h1>Study material</h1>
<div id="learn-functions-and-iteration" class="section level2">
<h2>Learn functions and iteration</h2>
<p>At this point, taking a moment to familiarize yourself with some R packages and programming concepts might be adviseable.</p>
<ol style="list-style-type: decimal">
<li>Learn to read and write your own functions in R</li>
<li>Learn to iterate procedures in R</li>
</ol>
<div id="write-functions-in-r" class="section level3">
<h3>Write functions in R</h3>
<p><a href="https://swcarpentry.github.io/r-novice-inflammation/02-func-R/" class="uri">https://swcarpentry.github.io/r-novice-inflammation/02-func-R/</a></p>
<p><a href="https://r4ds.had.co.nz/functions.html" class="uri">https://r4ds.had.co.nz/functions.html</a></p>
</div>
<div id="iteration-in-the-tidyverse" class="section level3">
<h3>Iteration in the tidyverse</h3>
<p>Packages <code>purrr</code> and (for parallel processing) <code>furrr</code>.</p>
<p><a href="https://r4ds.had.co.nz/iteration.html" class="uri">https://r4ds.had.co.nz/iteration.html</a></p>
<p><a href="https://www.r-bloggers.com/2021/09/tidy-parallel-processing-in-r-with-furrr/" class="uri">https://www.r-bloggers.com/2021/09/tidy-parallel-processing-in-r-with-furrr/</a></p>
<p><a href="https://purrr.tidyverse.org/" class="uri">https://purrr.tidyverse.org/</a></p>
<p><a href="https://dcl-prog.stanford.edu/purrr-mutate.html" class="uri">https://dcl-prog.stanford.edu/purrr-mutate.html</a></p>
<p><a href="https://dcl-prog.stanford.edu/purrr-parallel.html" class="uri">https://dcl-prog.stanford.edu/purrr-parallel.html</a></p>
<p><a href="https://furrr.futureverse.org/" class="uri">https://furrr.futureverse.org/</a></p>
</div>
</div>
<div id="learn-mixed-models" class="section level2">
<h2>Learn mixed-models</h2>
<p><a href="https://si.biostat.washington.edu/sites/default/files/modules/Seattle-SISG-18-IntroQG-Lecture08.pdf" class="uri">https://si.biostat.washington.edu/sites/default/files/modules/Seattle-SISG-18-IntroQG-Lecture08.pdf</a></p>
<p><a href="https://towardsdatascience.com/how-linear-mixed-model-works-350950a82911" class="uri">https://towardsdatascience.com/how-linear-mixed-model-works-350950a82911</a></p>
<p><a href="https://stats.oarc.ucla.edu/other/mult-pkg/introduction-to-linear-mixed-models/" class="uri">https://stats.oarc.ucla.edu/other/mult-pkg/introduction-to-linear-mixed-models/</a></p>
<p><a href="https://www.youtube.com/playlist?list=PLZ0lafzH_UmfHsoePj0t7xQglKSOpzbDp">Excellence in Breeding - Basics of GS - 2021 YouTube Video Series with Eduardo Covarrubias</a></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.1 (2021-08-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lme4_1.1-27.1            sommer_4.1.5             crayon_1.4.2            
 [4] lattice_0.20-45          MASS_7.3-54              Matrix_1.4-0            
 [7] genomicMateSelectR_0.2.0 forcats_0.5.1            stringr_1.4.0           
[10] dplyr_1.0.7              purrr_0.3.4              readr_2.1.1             
[13] tidyr_1.1.4              tibble_3.1.6             ggplot2_3.3.5           
[16] tidyverse_1.3.1          workflowr_1.7.0         

loaded via a namespace (and not attached):
 [1] httr_1.4.2       sass_0.4.0       jsonlite_1.7.2   splines_4.1.1   
 [5] here_1.0.1       modelr_0.1.8     bslib_0.3.1      assertthat_0.2.1
 [9] getPass_0.2-2    cellranger_1.1.0 yaml_2.2.1       pillar_1.6.4    
[13] backports_1.4.1  glue_1.6.0       digest_0.6.29    promises_1.2.0.1
[17] minqa_1.2.4      rvest_1.0.2      colorspace_2.0-2 htmltools_0.5.2 
[21] httpuv_1.6.5     pkgconfig_2.0.3  broom_0.7.11     haven_2.4.3     
[25] scales_1.1.1     processx_3.5.2   whisker_0.4      later_1.3.0     
[29] tzdb_0.2.0       git2r_0.29.0     generics_0.1.1   ellipsis_0.3.2  
[33] withr_2.4.3      cli_3.1.0        magrittr_2.0.1   readxl_1.3.1    
[37] evaluate_0.14    ps_1.6.0         fs_1.5.2         fansi_0.5.0     
[41] nlme_3.1-153     xml2_1.3.3       tools_4.1.1      hms_1.1.1       
[45] lifecycle_1.0.1  munsell_0.5.0    reprex_2.0.1     callr_3.7.0     
[49] compiler_4.1.1   jquerylib_0.1.4  rlang_0.4.12     nloptr_1.2.2.3  
[53] grid_4.1.1       rstudioapi_0.13  rmarkdown_2.11   boot_1.3-28     
[57] gtable_0.3.0     DBI_1.1.2        R6_2.5.1         lubridate_1.8.0 
[61] knitr_1.37       fastmap_1.1.0    utf8_1.2.2       rprojroot_2.0.2 
[65] stringi_1.7.6    Rcpp_1.0.7       vctrs_0.3.8      dbplyr_2.1.1    
[69] tidyselect_1.1.1 xfun_0.29       </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
