<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Home</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GSexample2022</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="download.html">Download</a>
</li>
<li>
  <a href="clean_phenos.html">Clean phenos</a>
</li>
<li>
  <a href="prep_genos.html">Prep genos</a>
</li>
<li>
  <a href="get_blups.html">Get BLUPs</a>
</li>
<li>
  <a href="intro_gp.html">Intro to GP</a>
</li>
<li>
  <a href="kfoldcrossval.html">K-fold Cross-val</a>
</li>
<li>
  <a href="get_gebv.html">Get GEBVs</a>
</li>
<li>
  <a href="parentwise_cv.html">Parent-wise CV</a>
</li>
<li>
  <a href="pred_crosses.html">Predict crosses</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Home</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-05-09
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>GSexample2022/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220320code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20220320)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220320code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220320)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomwolfemdGSexample2022treebb2d98179db87acbe2189a400bac5d598f9119d8targetblankbb2d981a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/wolfemd/GSexample2022/tree/bb2d98179db87acbe2189a400bac5d598f9119d8" target="_blank">bb2d981</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomwolfemdGSexample2022treebb2d98179db87acbe2189a400bac5d598f9119d8targetblankbb2d981a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/wolfemd/GSexample2022/tree/bb2d98179db87acbe2189a400bac5d598f9119d8" target="_blank">bb2d981</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  .DS_Store
    Untracked:  Icon
    Untracked:  analysis/.DS_Store
    Untracked:  analysis/BreedBaseGenotypes_subset.log
    Untracked:  analysis/Icon
    Untracked:  analysis/images/Icon
    Untracked:  analysis/images/paste-2F34C15B.png
    Untracked:  analysis/images/paste-AD7AEEAC.png
    Untracked:  analysis/images/paste-CB5D91AC.png
    Untracked:  analysis/troubleshooting.Rmd
    Untracked:  code/Icon
    Untracked:  data/.DS_Store
    Untracked:  data/BreedBaseGenotypesDownload.positions
    Untracked:  data/BreedBaseGenotypesDownload.vcf
    Untracked:  data/BreedBaseGenotypes_subset.bed
    Untracked:  data/BreedBaseGenotypes_subset.bim
    Untracked:  data/BreedBaseGenotypes_subset.fam
    Untracked:  data/BreedBaseGenotypes_subset.hap.gz
    Untracked:  data/BreedBaseGenotypes_subset.log
    Untracked:  data/BreedBaseGenotypes_subset.nosex
    Untracked:  data/BreedBaseGenotypes_subset.positions
    Untracked:  data/BreedBaseGenotypes_subset.samples
    Untracked:  data/Icon
    Untracked:  data/metadata_cynthia.csv
    Untracked:  data/phenotype_cynthia.csv
    Untracked:  data/subset_unique_names_for_vcf.txt
    Untracked:  data/unique_names_for_vcf.txt
    Untracked:  data/vcf_colnames.txt
    Untracked:  out.log
    Untracked:  output/.DS_Store
    Untracked:  output/BreedBaseGenotypes_subset.genome
    Untracked:  output/BreedBaseGenotypes_subset.log
    Untracked:  output/BreedBaseGenotypes_subset.nosex
    Untracked:  output/Icon
    Untracked:  output/genomicPredictions.rds
    Untracked:  output/models_compared.Rdata
    Untracked:  output/parentWiseCV.rds
    Untracked:  output/standardCV.rds
    Untracked:  output/verified_ped.rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/index.Rmd</code>) and HTML (<code>docs/index.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/bb2d98179db87acbe2189a400bac5d598f9119d8/analysis/index.Rmd" target="_blank">bb2d981</a>
</td>
<td>
wolfemd
</td>
<td>
2022-05-09
</td>
<td>
Add the final two sections - parent-wise cross-validation plus predicting crosses
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/GSexample2022/260d2fed9c49fc20aa3f1b2b41d5bfa2cd99aecc/docs/index.html" target="_blank">260d2fe</a>
</td>
<td>
wolfemd
</td>
<td>
2022-05-05
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/975daa79a6e60fffa81b189ad3803dbd080b9ba4/analysis/index.Rmd" target="_blank">975daa7</a>
</td>
<td>
wolfemd
</td>
<td>
2022-05-05
</td>
<td>
Publish demo for k-fold cross-val and genomic prediction of GEBV steps
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/GSexample2022/176b527de592e875f34a446c60c765f327c67eb6/docs/index.html" target="_blank">176b527</a>
</td>
<td>
wolfemd
</td>
<td>
2022-04-26
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/acd7ed55418e569a295cd90a5404e7e8d8133bdb/analysis/index.Rmd" target="_blank">acd7ed5</a>
</td>
<td>
wolfemd
</td>
<td>
2022-04-26
</td>
<td>
Add ‘Intro to GP’ section. Update ‘Index’ page. Minor tweak to ‘Get BLUPs’ section.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/GSexample2022/63d6587f701fd0af1c9e5c0c2a97caf8f6273447/docs/index.html" target="_blank">63d6587</a>
</td>
<td>
wolfemd
</td>
<td>
2022-04-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/9c0e19e38ee5c830efc736987c5fd8e9e6450616/analysis/index.Rmd" target="_blank">9c0e19e</a>
</td>
<td>
wolfemd
</td>
<td>
2022-04-20
</td>
<td>
Add the ‘preliminary trial analysis’ or ‘get blups’ steps.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/GSexample2022/065ed11e011658a7627dacec5d72604779417326/docs/index.html" target="_blank">065ed11</a>
</td>
<td>
wolfemd
</td>
<td>
2022-04-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/e3d8e656fa9569396dbd5ea381bc87edf99da39d/analysis/index.Rmd" target="_blank">e3d8e65</a>
</td>
<td>
wolfemd
</td>
<td>
2022-04-01
</td>
<td>
update everything
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/GSexample2022/adb7ede079281ff2c2bf22c710b3e2c7d2cccc55/docs/index.html" target="_blank">adb7ede</a>
</td>
<td>
wolfemd
</td>
<td>
2022-04-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/b62cd6d4b3a93bcc0db9fb11e0366c9d6b311fdb/analysis/index.Rmd" target="_blank">b62cd6d</a>
</td>
<td>
wolfemd
</td>
<td>
2022-04-01
</td>
<td>
Add the ‘prep genos’ steps
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/GSexample2022/4d5fb817a885621aa8c263f74cfa5f0ceb94b787/docs/index.html" target="_blank">4d5fb81</a>
</td>
<td>
wolfemd
</td>
<td>
2022-03-25
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/11eeb7936bf59e7438091a822bb16f38d42bce44/analysis/index.Rmd" target="_blank">11eeb79</a>
</td>
<td>
wolfemd
</td>
<td>
2022-03-25
</td>
<td>
Publish cleaned phenotypes.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/5d8319287f14dbf3f645a0622eca1c1f5c7d53af/analysis/index.Rmd" target="_blank">5d83192</a>
</td>
<td>
wolfemd
</td>
<td>
2022-03-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/GSexample2022/5d8319287f14dbf3f645a0622eca1c1f5c7d53af/docs/index.html" target="_blank">5d83192</a>
</td>
<td>
wolfemd
</td>
<td>
2022-03-24
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/4665bfec1a68801e7d38a5dfda5ad13dece90206/analysis/index.Rmd" target="_blank">4665bfe</a>
</td>
<td>
wolfemd
</td>
<td>
2022-03-21
</td>
<td>
fix link to ‘downloads.html’
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/GSexample2022/4665bfec1a68801e7d38a5dfda5ad13dece90206/docs/index.html" target="_blank">4665bfe</a>
</td>
<td>
wolfemd
</td>
<td>
2022-03-21
</td>
<td>
fix link to ‘downloads.html’
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/GSexample2022/a2e74926c8651a089b12d7f94f65e46251125e42/docs/index.html" target="_blank">a2e7492</a>
</td>
<td>
wolfemd
</td>
<td>
2022-03-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/e4f9fd907dcedd581b8977a735ac209be218f44d/analysis/index.Rmd" target="_blank">e4f9fd9</a>
</td>
<td>
wolfemd
</td>
<td>
2022-03-20
</td>
<td>
Update index.Rmd with additional screenshipts on publishing to github
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/GSexample2022/9a87a77cd915c5b7ce6ea38cd1d3041536ca3918/docs/index.html" target="_blank">9a87a77</a>
</td>
<td>
wolfemd
</td>
<td>
2022-03-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/wolfemd/GSexample2022/b54d21a7eae89be6fb7567265e15f7c9bc0a8f0e/docs/index.html" target="_blank">b54d21a</a>
</td>
<td>
wolfemd
</td>
<td>
2022-03-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/2a55fb9d0b905a347b0456a6b81f4c9db6569bd7/analysis/index.Rmd" target="_blank">2a55fb9</a>
</td>
<td>
wolfemd
</td>
<td>
2022-03-20
</td>
<td>
First publish
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/3366366bca152906150c06073f78c359a45ac3d3/analysis/index.Rmd" target="_blank">3366366</a>
</td>
<td>
wolfemd
</td>
<td>
2022-03-20
</td>
<td>
First commit. Index + Download page. Include images and phenotype/metadata files for workshop purposes only!
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/d05e5dd638d2a2e2930615b2529b752bfe5111f1/analysis/index.Rmd" target="_blank">d05e5dd</a>
</td>
<td>
wolfemd
</td>
<td>
2022-03-20
</td>
<td>
Start workflowr project.
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>Welcome to my example GS repository. I followed the instructions <a href="https://wolfemd.github.io/GenomicSelectionManual/create_project.html">here</a> to create this project.</p>
<p>I’m going to follow along the <a href="https://wolfemd.github.io/GenomicSelectionManual/">Genomic Selection Manual</a> throughout this project.</p>
<div id="data-download" class="section level2">
<h2>Data download</h2>
<p>Create the “download.Rmd”</p>
<pre class="r"><code>wflow_open(&quot;analysis/download.Rmd&quot;)</code></pre>
<p>Link to it: <a href="download.html">Data Download Stage</a></p>
<p>Add it to the top navigation bar, via the <code>_site.yml</code> file:</p>
<p><img src="images/paste-72D2779D.png" /></p>
</div>
<div id="table-of-contents" class="section level2">
<h2>Table of Contents</h2>
<ol style="list-style-type: decimal">
<li><a href="download.html">Data Download Stage</a></li>
<li><a href="clean_phenos.html">Clean phenotype data</a>: standard procedures to prepare phenotypes for analysis.</li>
<li><a href="prep_genos.html">Prepare genotypic data</a>: standard procedures to prepare genotypic inputs (e.g. haplotype matrices) for genomic mate selection analyses.</li>
<li><a href="get_blups.html">Preliminary field trial analyses</a>: Conduct preliminary analysis of the trial data without genomic relatedeness / marker-information. Identify the best-fitting model for the data and potentially curate the raw data, removing outliers. Fit mixed-model to MET data. Extract BLUPs, PEVs and variance components (VarComps). Compute de-regressed BLUPs (drgBLUPs) and weights (WTS) for “Stage 2.”</li>
<li><a href="intro_gp.html">Introduction to Genomic Prediction in R</a>: Into to using <code>sommer</code> to fit G-BLUP and RR-BLUP models, a demo of their equivalency and quickly showing how to run cross-validations with the <code>runCrossVal()</code> function of <code>library(genomicMateSelectR)</code>.</li>
<li><a href="kfoldcrossval.html">Check prediction accuracy</a>: Use what we learned previously to run a k-fold cross-validation and estimate the accuracy predicting the performance of parents on our selection index.</li>
<li><a href="get_gebv.html">Predict parental GEBV</a>: Having checked prediction accuracy and assuming we are satisfied, we can now run the full genomic prediction using <code>runGenomicPredictions()</code> and get GEBV for our population, computing the selection index performance using those GEBVs for our ultimate selections of parents.</li>
<li><a href="parentwise_cv.html">Parent-wise Cross-validation</a>: Check the accuracy predicting the means, variances and covariances in previously untested crosses. <em>Also</em> verify using marker data that the pedigree is correct, before doing the parent-wise CV.</li>
<li><a href="pred_crosses.html">Predict crosses</a>: The main event. Predict the means, variances and usefulness on the selection index and select the top crosses.</li>
</ol>
<p><strong>TROUBLESHOOTING RMD:</strong> <a href="troubleshooting.html">Troubleshooting each step</a></p>
</div>
<div id="committing-and-publishing" class="section level2">
<h2>Committing and publishing</h2>
<p>Example, first commit:</p>
<p><img src="images/paste-1EE3E624.png" /></p>
<p><strong>VERY IMPORTANT NOTE:</strong> here I’m adding the <code>data/metadata.csv</code> and <code>data/phenotype.csv</code>. These are small files and I want them available to workshop users. DO NOT make a habit of this. GitHub max size is 25 Mb. ESPECIALLY DO NOT add genomic data files to your Git version control.</p>
<p><img src="images/paste-90E446A0.png" /></p>
</div>
<div id="publishing-to-github-pages" class="section level2">
<h2>Publishing to GitHub Pages</h2>
<p>Follow the manual’s instructions at: <a href="https://wolfemd.github.io/GenomicSelectionManual/create_project.html#publishing-on-github-pages" class="uri">https://wolfemd.github.io/GenomicSelectionManual/create_project.html#publishing-on-github-pages</a></p>
<p>In addition, edit repository details, adding the GitHub pages web address and a short description, like so:</p>
<p><img src="images/paste-CB5D91AC.png" /></p>
<p><img src="images/paste-AD7AEEAC.png" /></p>
<p>Consider modifying the <code>README.md</code> file in the main project folder, as it will display to anyone viewing the repository directly on GitHub.</p>
<p><img src="images/paste-2F34C15B.png" /></p>
</div>
<div id="customize-workflowr-project" class="section level2">
<h2>Customize <code>workflowR</code> project</h2>
<div id="inline-chunk-output" class="section level3">
<h3>Inline chunk output</h3>
<p>Change chunk output setting to “inline output” instead of “console”</p>
<p><img src="images/paste-46ECDEBA.png" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.1 (2021-08-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] workflowr_1.7.0

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.7       compiler_4.1.1   pillar_1.6.4     bslib_0.3.1     
 [5] later_1.3.0      git2r_0.29.0     jquerylib_0.1.4  tools_4.1.1     
 [9] getPass_0.2-2    digest_0.6.29    jsonlite_1.7.2   evaluate_0.14   
[13] tibble_3.1.6     lifecycle_1.0.1  pkgconfig_2.0.3  rlang_0.4.12    
[17] rstudioapi_0.13  yaml_2.2.1       xfun_0.29        fastmap_1.1.0   
[21] httr_1.4.2       stringr_1.4.0    knitr_1.37       sass_0.4.0      
[25] fs_1.5.2         vctrs_0.3.8      rprojroot_2.0.2  glue_1.6.0      
[29] R6_2.5.1         processx_3.5.2   fansi_0.5.0      rmarkdown_2.11  
[33] callr_3.7.0      magrittr_2.0.1   whisker_0.4      ps_1.6.0        
[37] promises_1.2.0.1 htmltools_0.5.2  ellipsis_0.3.2   httpuv_1.6.5    
[41] utf8_1.2.2       stringi_1.7.6    crayon_1.4.2    </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
