<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="wolfemd" />

<meta name="date" content="2022-04-25" />

<title>Introduction to Genomic Prediction in R</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GSexample2022</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="download.html">Download</a>
</li>
<li>
  <a href="clean_phenos.html">Clean phenos</a>
</li>
<li>
  <a href="prep_genos.html">Prep genos</a>
</li>
<li>
  <a href="get_blups.html">Get BLUPs</a>
</li>
<li>
  <a href="intro_gp.html">Intro to GP</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Introduction to Genomic Prediction in R</h1>
<h4 class="author">wolfemd</h4>
<h4 class="date">2022-04-25</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-04-26
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>GSexample2022/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220320code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20220320)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220320code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220320)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomwolfemdGSexample2022treeacd7ed55418e569a295cd90a5404e7e8d8133bdbtargetblankacd7ed5a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/wolfemd/GSexample2022/tree/acd7ed55418e569a295cd90a5404e7e8d8133bdb" target="_blank">acd7ed5</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomwolfemdGSexample2022treeacd7ed55418e569a295cd90a5404e7e8d8133bdbtargetblankacd7ed5a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/wolfemd/GSexample2022/tree/acd7ed55418e569a295cd90a5404e7e8d8133bdb" target="_blank">acd7ed5</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  .DS_Store
    Untracked:  Icon
    Untracked:  analysis/.DS_Store
    Untracked:  analysis/Icon
    Untracked:  analysis/images/Icon
    Untracked:  analysis/images/paste-2F34C15B.png
    Untracked:  analysis/images/paste-AD7AEEAC.png
    Untracked:  analysis/images/paste-CB5D91AC.png
    Untracked:  analysis/troubleshooting.Rmd
    Untracked:  code/Icon
    Untracked:  data/.DS_Store
    Untracked:  data/BreedBaseGenotypesDownload.positions
    Untracked:  data/BreedBaseGenotypesDownload.vcf
    Untracked:  data/BreedBaseGenotypes_subset.hap.gz
    Untracked:  data/BreedBaseGenotypes_subset.positions
    Untracked:  data/BreedBaseGenotypes_subset.samples
    Untracked:  data/Icon
    Untracked:  data/metadata_cynthia.csv
    Untracked:  data/phenotype_cynthia.csv
    Untracked:  data/subset_unique_names_for_vcf.txt
    Untracked:  data/unique_names_for_vcf.txt
    Untracked:  data/vcf_colnames.txt
    Untracked:  out.log
    Untracked:  output/.DS_Store
    Untracked:  output/Icon
    Untracked:  output/models_compared.Rdata

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/intro_gp.Rmd</code>) and HTML (<code>docs/intro_gp.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/acd7ed55418e569a295cd90a5404e7e8d8133bdb/analysis/intro_gp.Rmd" target="_blank">acd7ed5</a>
</td>
<td>
wolfemd
</td>
<td>
2022-04-26
</td>
<td>
Add ‘Intro to GP’ section. Update ‘Index’ page. Minor tweak to ‘Get BLUPs’ section.
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="load-packages" class="section level1">
<h1>Load packages</h1>
<pre class="r"><code>suppressMessages(library(tidyverse)); 
suppressMessages(library(genomicMateSelectR)); 
suppressMessages(library(sommer));</code></pre>
</div>
<div id="kinship-matrix" class="section level1">
<h1>Kinship matrix</h1>
<pre class="r"><code>A&lt;-readRDS(file=here::here(&quot;output&quot;,&quot;kinship_add.rds&quot;))</code></pre>
<pre class="r"><code>dim(A)</code></pre>
<pre><code>[1] 1007 1007</code></pre>
<pre class="r"><code>hist(A[lower.tri(A)], main=&#39;Off-diagonal values (lower-triangle) of the kinship matrix&#39;)</code></pre>
<p><img src="figure/intro_gp.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /> The off-diagonals express the genomic relatedness between different individuals in the dataset.</p>
<p>This formulation has a mean-value of zero, such that &gt;0 kinship means “more than the average population level of relatedness” while &lt;0 values mean “greater than average.”</p>
<pre class="r"><code>hist(diag(A), main=&#39;Diagonals of the kinship&#39;, breaks=20)</code></pre>
<p><img src="figure/intro_gp.Rmd/unnamed-chunk-5-1.png" width="672" style="display: block; margin: auto;" /> The diagonal values express the relatedness of each individual to itself, which should be interpreted as the level of inbreeding (homozygosity).</p>
<pre class="r"><code>heatmap(A)</code></pre>
<p><img src="figure/intro_gp.Rmd/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="gblup-prediction" class="section level1">
<h1>GBLUP prediction</h1>
<pre class="r"><code>blups&lt;-readRDS(here::here(&quot;output&quot;,&quot;blups.rds&quot;))</code></pre>
<pre class="r"><code># pull out one trait (DM) BLUPs
dm_blups&lt;-blups$blups[[1]]
dm_blups %&gt;% head</code></pre>
<pre><code># A tibble: 6 × 6
  germplasmName    BLUP   PEV   REL drgBLUP    WT
  &lt;chr&gt;           &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;
1 50395         -1.02    4.44 0.117  -8.65  0.177
2 95NA-00063    -0.0285  4.44 0.117  -0.243 0.177
3 ANKRA         -0.760   4.44 0.117  -6.47  0.177
4 BAHKYEHEMAA    0.924   4.44 0.117   7.87  0.177
5 BANGWEULU      0.331   4.44 0.117   2.82  0.177
6 BOUNOUA_BLANC  1.02    4.44 0.117   8.68  0.177</code></pre>
<p>Now we have all we need to implement a GBLUP model using mmer().</p>
<p>One thing to note:</p>
<pre class="r"><code>table(dm_blups$germplasmName %in% rownames(A))</code></pre>
<pre><code>
FALSE  TRUE 
  564   962 </code></pre>
<p>There are phenotyped-but-not-genotyped lines.</p>
<pre class="r"><code>table(rownames(A) %in% dm_blups$germplasmName)</code></pre>
<pre><code>
FALSE  TRUE 
   45   962 </code></pre>
<p>There are also genotyped-but-not-phenotyped lines.</p>
<p>Phenotyped-but-not-genotyped lines need to be excluded from the analysis.</p>
<p>Genotyped-but-not-phenotyped lines, as discussed above are accepted and can be predicted!</p>
<pre class="r"><code>dm_blups %&lt;&gt;% 
     filter(germplasmName %in% rownames(A))</code></pre>
<pre class="r"><code>gblup_all&lt;-mmer(fixed = drgBLUP~1,
                # here we specify a random-effect for the &quot;germplasmName&quot; variable
                # and supply the kinship matrix &quot;A&quot; as follows:
                random = ~vs(germplasmName,Gu=A),
                weights = WT,
                data=dm_blups)</code></pre>
<pre><code>Version out of date. Please update sommer to the newest version using:
install.packages(&#39;sommer&#39;) in a new session
 Use the &#39;date.warning&#39; argument to disable the warning message.Adding additional levels of Gu in the model matrix of &#39;germplasmName&#39; 
iteration    LogLik     wall    cpu(sec)   restrained
    1      -356.471   6:52:30      3           0
    2      -355.724   6:52:33      6           0
    3      -355.405   6:52:35      8           0
    4      -355.333   6:52:38      11           0
    5      -355.327   6:52:41      14           0
    6      -355.326   6:52:43      16           0</code></pre>
<p>Here’s how to get out the <strong>GEBV</strong> from the <code>mmer()</code> output:</p>
<pre class="r"><code>gebv&lt;-gblup_all$U$`u:germplasmName`$drgBLUP
# Notice the &quot;germplasmName&quot; and &quot;drgBLUP&quot; references in this call 
## are specific to the analysis/dataset in question. 
## See the sommer manual on CRAN. </code></pre>
<p>Things to notice:</p>
<pre class="r"><code>length(gebv)</code></pre>
<pre><code>[1] 1007</code></pre>
<p>Only 962 phenotyped lines.</p>
<p>1007 predicted GEBV.</p>
<pre class="r"><code>summary(gebv)</code></pre>
<pre><code>    Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
-4.76399 -0.63157  0.05915  0.00000  0.68340  2.44024 </code></pre>
<p>GEBV are centered on zero.</p>
<pre class="r"><code>hist(gebv)</code></pre>
<p><img src="figure/intro_gp.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="prediction-accuracy" class="section level1">
<h1>Prediction accuracy</h1>
<p>Now we can move into a more potentially complicated arena: evaluating prediction accuracy.</p>
<p>There are many scenarios and styles for doing this.</p>
<p>Let’s start <em>super</em> simple.</p>
<p>Of the phenotyped lines, I randomly sample approx. 1/5 (20%).</p>
<pre class="r"><code>set.seed(1212)
test_set&lt;-sample(dm_blups$germplasmName,size = ceiling(346/5), replace = F)</code></pre>
<p>Remove the lines chosen (<strong>test_set</strong>) from the training dataset:</p>
<pre class="r"><code>training_blups&lt;-dm_blups %&gt;% 
     filter(!germplasmName %in% test_set)</code></pre>
<p>Fit the prediction model:</p>
<pre class="r"><code>gblup_train&lt;-mmer(fixed = drgBLUP~1,
                  # here we specify a random-effect for the &quot;germplasmName&quot; variable
                  # and supply the kinship matrix &quot;A&quot; as follows:
                  random = ~vs(germplasmName,Gu=A),
                  weights = WT,
                  data=training_blups)</code></pre>
<pre><code>Version out of date. Please update sommer to the newest version using:
install.packages(&#39;sommer&#39;) in a new session
 Use the &#39;date.warning&#39; argument to disable the warning message.Adding additional levels of Gu in the model matrix of &#39;germplasmName&#39; 
iteration    LogLik     wall    cpu(sec)   restrained
    1      -337.163   6:52:49      2           0
    2      -336.308   6:52:51      4           0
    3      -335.959   6:52:53      6           0
    4      -335.885   6:52:55      8           0
    5      -335.879   6:52:57      10           0
    6      -335.878   6:52:59      12           0</code></pre>
<p>Now we can compare (correlate) the GEBV predicted for the held-out <strong>test_set</strong> lines to the BLUPs for those lines.</p>
<p>This is considered an estimate of <strong>“prediction accuracy”</strong>.</p>
<pre class="r"><code># grab the gebv only for the test_set lines
gebv_test&lt;-gblup_train$U$`u:germplasmName`$drgBLUP[test_set]
# merge the original BLUPs (for the test set only) to the corresponding GEBV
# I use a left_join() or merge() to be absolutely sure there are no mix-ups
gebv_vs_blups_testset&lt;-dm_blups %&gt;% 
     filter(germplasmName %in% test_set) %&gt;% 
     left_join(tibble(germplasmName=names(gebv_test),GEBV=as.numeric(gebv_test)))</code></pre>
<pre><code>Joining, by = &quot;germplasmName&quot;</code></pre>
<p>Correlate the GEBV with the BLUP for the test_set.</p>
<pre class="r"><code>gebv_vs_blups_testset %$% cor(BLUP,GEBV)</code></pre>
<pre><code>[1] 0.1750126</code></pre>
<p>This correlation represents an estimate of accuracy predicting lines that aren’t phenotyped, based only on their genetic relatedness and the phenotypes of their relatives.</p>
<p>Sure, this is a low values. But it’s a small dataset using a small sample of markers to make compute fast for example purposes.</p>
</div>
<div id="cross-validation" class="section level1">
<h1>Cross-validation</h1>
<p>Cross-validation functions get complicated quickly as they have loops within loops.</p>
<p>I built a function <code>runCrossVal()</code> that fits several different models, handles multiple traits, can run in-parallel across mutiple compute-cores and even accepts selection index weights to compute selection index accuracy. <code>runCrossVal()</code> is included in the <code>genomicMateSelectR</code> package.</p>
<p>Below, I use it on only 1 trait, additive-only model.</p>
<p><strong>NOTES:</strong> - doesn’t work properly <em>unless</em> <code>gid="GID"</code>. - It also requires you to remove any phenotyped lines that aren’t genotyped.</p>
<p>Following chunk does both of those:</p>
<pre class="r"><code>blups_forRunCrossValFunc&lt;-blups %&gt;% 
     # just one trait
     slice(1) %&gt;% 
     # need to rename the &quot;blups&quot; list to comply with the runCrossVal function
     rename(TrainingData=blups) %&gt;% 
     dplyr::select(Trait,TrainingData) %&gt;% 
     # need also to remove phenotyped-but-not-genotyped lines
     # couldn&#39;t hurt to also subset the kinship to only phenotyped lines... would save RAM
     mutate(TrainingData=map(TrainingData,
                             ~filter(.,germplasmName %in% rownames(A)) %&gt;% 
                                  # rename the germplasmName column to GID
                                  rename(GID=germplasmName)))</code></pre>
<pre class="r"><code>standardCV&lt;-runCrossVal(blups=blups_forRunCrossValFunc,
                        modelType=&quot;A&quot;,
                        selInd=FALSE,
                        grms=list(A=A),
                        nrepeats=2,nfolds=5,
                        gid=&quot;GID&quot;,seed=424242,
                        ncores=5)</code></pre>
<pre><code>Loading required package: rsample</code></pre>
<pre><code>Loading required package: furrr</code></pre>
<pre><code>Loading required package: future</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -282.488   6:53:6      2           0
    2      -281.563   6:53:8      4           0
    3      -281.236   6:53:9      5           0
    4      -281.183   6:53:11      7           0
    5      -281.18   6:53:12      8           0
    6      -281.18   6:53:14      10           0
[1] &quot;GBLUP model complete - one trait&quot;
[1] &quot;Genomic predictions done for all traits in one repeat-fold&quot;
iteration    LogLik     wall    cpu(sec)   restrained
    1      -278.563   6:53:18      2           0
    2      -277.344   6:53:20      4           0
    3      -276.812   6:53:21      5           0
    4      -276.682   6:53:23      7           0
    5      -276.668   6:53:24      8           0
    6      -276.666   6:53:26      10           0
    7      -276.666   6:53:27      11           0
[1] &quot;GBLUP model complete - one trait&quot;
[1] &quot;Genomic predictions done for all traits in one repeat-fold&quot;</code></pre>
<pre><code>Joining, by = &quot;GID&quot;</code></pre>
<pre><code>Joining, by = &quot;GID&quot;</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -284.947   6:53:7      2           0
    2      -283.172   6:53:8      3           0
    3      -282.401   6:53:10      5           0
    4      -282.225   6:53:11      6           0
    5      -282.21   6:53:13      8           0
    6      -282.209   6:53:15      10           0
    7      -282.209   6:53:17      12           0
[1] &quot;GBLUP model complete - one trait&quot;
[1] &quot;Genomic predictions done for all traits in one repeat-fold&quot;
iteration    LogLik     wall    cpu(sec)   restrained
    1      -285.414   6:53:20      2           0
    2      -285.212   6:53:22      4           0
    3      -285.127   6:53:24      6           0
    4      -285.109   6:53:25      7           0
    5      -285.108   6:53:27      9           0
    6      -285.108   6:53:28      10           0
[1] &quot;GBLUP model complete - one trait&quot;
[1] &quot;Genomic predictions done for all traits in one repeat-fold&quot;</code></pre>
<pre><code>Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -293.434   6:53:7      2           0
    2      -292.797   6:53:9      4           0
    3      -292.531   6:53:10      5           0
    4      -292.475   6:53:12      7           0
    5      -292.471   6:53:14      9           0
    6      -292.47   6:53:16      11           0
[1] &quot;GBLUP model complete - one trait&quot;
[1] &quot;Genomic predictions done for all traits in one repeat-fold&quot;
iteration    LogLik     wall    cpu(sec)   restrained
    1      -299.215   6:53:19      2           0
    2      -298.959   6:53:21      4           0
    3      -298.843   6:53:22      5           0
    4      -298.815   6:53:24      7           0
    5      -298.812   6:53:26      9           0
    6      -298.811   6:53:27      10           0
[1] &quot;GBLUP model complete - one trait&quot;
[1] &quot;Genomic predictions done for all traits in one repeat-fold&quot;</code></pre>
<pre><code>Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -270.158   6:53:8      2           0
    2      -270.127   6:53:9      3           0
    3      -270.114   6:53:11      5           0
    4      -270.111   6:53:13      7           0
    5      -270.111   6:53:14      8           0
[1] &quot;GBLUP model complete - one trait&quot;
[1] &quot;Genomic predictions done for all traits in one repeat-fold&quot;
iteration    LogLik     wall    cpu(sec)   restrained
    1      -280.573   6:53:18      2           0
    2      -278.518   6:53:20      4           0
    3      -277.652   6:53:21      5           0
    4      -277.453   6:53:23      7           0
    5      -277.435   6:53:25      9           0
    6      -277.433   6:53:26      10           0
    7      -277.433   6:53:28      12           0
[1] &quot;GBLUP model complete - one trait&quot;
[1] &quot;Genomic predictions done for all traits in one repeat-fold&quot;</code></pre>
<pre><code>Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -288.833   6:53:8      2           0
    2      -287.363   6:53:10      4           0
    3      -286.82   6:53:12      6           0
    4      -286.726   6:53:13      7           0
    5      -286.721   6:53:15      9           0
    6      -286.721   6:53:17      11           0
[1] &quot;GBLUP model complete - one trait&quot;
[1] &quot;Genomic predictions done for all traits in one repeat-fold&quot;
iteration    LogLik     wall    cpu(sec)   restrained
    1      -287.812   6:53:20      2           0
    2      -286.737   6:53:22      4           0
    3      -286.284   6:53:24      6           0
    4      -286.189   6:53:25      7           0
    5      -286.182   6:53:27      9           0
    6      -286.181   6:53:28      10           0
[1] &quot;GBLUP model complete - one trait&quot;
[1] &quot;Genomic predictions done for all traits in one repeat-fold&quot;</code></pre>
<pre><code>Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;</code></pre>
<pre class="r"><code>standardCV %&gt;% unnest(accuracyEstOut) </code></pre>
<pre><code># A tibble: 10 × 9
   repeats  seeds splits  id    Trait predOf predVSobs Accuracy NcompleteTestPa…
     &lt;int&gt;  &lt;int&gt; &lt;list&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;  &lt;list&gt;       &lt;dbl&gt;            &lt;dbl&gt;
 1       1 395601 &lt;split… Fold1 DM    GEBV   &lt;tibble … 0.0983                  1
 2       1 395601 &lt;split… Fold2 DM    GEBV   &lt;tibble … 0.0437                  1
 3       1 395601 &lt;split… Fold3 DM    GEBV   &lt;tibble … 0.0905                  1
 4       1 395601 &lt;split… Fold4 DM    GEBV   &lt;tibble … 0.222                   1
 5       1 395601 &lt;split… Fold5 DM    GEBV   &lt;tibble … 0.144                   1
 6       2 215870 &lt;split… Fold1 DM    GEBV   &lt;tibble … 0.275                   1
 7       2 215870 &lt;split… Fold2 DM    GEBV   &lt;tibble … 0.196                   1
 8       2 215870 &lt;split… Fold3 DM    GEBV   &lt;tibble … 0.000840                1
 9       2 215870 &lt;split… Fold4 DM    GEBV   &lt;tibble … 0.126                   1
10       2 215870 &lt;split… Fold5 DM    GEBV   &lt;tibble … 0.139                   1</code></pre>
<pre class="r"><code>standardCV %&gt;% 
     unnest(accuracyEstOut) %$% summary(Accuracy)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
0.00084 0.09243 0.13274 0.13366 0.18305 0.27524 </code></pre>
</div>
<div id="rrblup-prediction" class="section level1">
<h1>RRBLUP prediction</h1>
<p>AKA SNP-BLUP.</p>
<p>Let’s see how to fit with <code>mmer()</code>.</p>
<p>Also demo the equivalency of RRBLUP and GBLUP.</p>
<pre class="r"><code># load the dosage matrix
M&lt;-readRDS(file=here::here(&quot;data&quot;,&quot;dosages.rds&quot;))
# re-load and extract the blups for just 1 trait
blups&lt;-readRDS(file=here::here(&quot;output&quot;,&quot;blups.rds&quot;))
dm_blups&lt;-blups$blups[[1]]
# remove phenotyped-but-not-genoypted blups
dm_blups %&lt;&gt;% 
     filter(germplasmName %in% rownames(M))
# remove genotyped-but-not-phenoypted (not strictly necessary)
M&lt;-M[rownames(M) %in% dm_blups$germplasmName,]
# Create a kinship matrix &quot;K&quot;
K&lt;-kinship(M,&quot;add&quot;)
# Now create a centered dosage matrix, to use as predictors for rr-BLUP
Z&lt;-centerDosage(M)</code></pre>
<p>Fit the GBLUP model</p>
<pre class="r"><code>gblup&lt;-mmer(fixed = drgBLUP~1,
            random = ~vs(germplasmName,Gu = K),
            weights = WT,
            data=dm_blups)</code></pre>
<pre><code>Version out of date. Please update sommer to the newest version using:
install.packages(&#39;sommer&#39;) in a new session
 Use the &#39;date.warning&#39; argument to disable the warning message.iteration    LogLik     wall    cpu(sec)   restrained
    1      -356.474   6:53:35      3           0
    2      -355.725   6:53:38      6           0
    3      -355.405   6:53:41      9           0
    4      -355.333   6:53:43      11           0
    5      -355.327   6:53:46      14           0
    6      -355.326   6:53:49      17           0</code></pre>
<pre class="r"><code># extract the G-BLUPs (GEBVs)
g&lt;-gblup$U$`u:germplasmName`$drgBLUP
all(names(g) %in% rownames(Z))</code></pre>
<pre><code>[1] TRUE</code></pre>
<p>Now fit the RRBLUP model</p>
<pre class="r"><code># one catch
# the rows of the centered dosage matrix must be matched
## to the rows of data.frame with the blups
## one way to do this is by creating and incidence matrix
## use the model.matrix() function as follows:
ZincMat=model.matrix(~factor(germplasmName,levels=rownames(Z))-1,data=dm_blups) # -1 because we don&#39;t want an intercept here
dim(ZincMat) # [1] 346 346 </code></pre>
<pre><code>[1] 962 962</code></pre>
<pre class="r"><code># relates the rows of the BLUPs to the rows the marker matrix

# now fit the RRBLUP model
rrblup&lt;-mmer(fixed = drgBLUP~1,
             random = ~vs(list(ZincMat%*%Z),buildGu = FALSE),
             weights = WT,
             data=dm_blups)</code></pre>
<pre><code>Version out of date. Please update sommer to the newest version using:
install.packages(&#39;sommer&#39;) in a new session
 Use the &#39;date.warning&#39; argument to disable the warning message.iteration    LogLik     wall    cpu(sec)   restrained
    1      -835.075   6:54:11      18           0
    2      -610.966   6:54:14      21           0
    3      -438.994   6:54:17      24           0
    4      -362.751   6:54:20      27           0
    5      -355.985   6:54:22      29           0
    6      -355.386   6:54:25      32           0
    7      -355.332   6:54:27      34           0
    8      -355.327   6:54:30      37           0
    9      -355.326   6:54:32      39           0</code></pre>
<p>Extract the SNP-BLUPs</p>
<pre class="r"><code># extract the SNP-BLUPs
u&lt;-as.matrix(rrblup$U$`u:Z`$drgBLUP)
all(rownames(u)==colnames(Z))</code></pre>
<pre><code>[1] TRUE</code></pre>
<pre class="r"><code>summary(u)</code></pre>
<pre><code>       V1            
 Min.   :-0.0311133  
 1st Qu.:-0.0043459  
 Median :-0.0003939  
 Mean   :-0.0004874  
 3rd Qu.: 0.0036510  
 Max.   : 0.0288460  </code></pre>
<pre class="r"><code>hist(u,breaks=20,main = &#39;Distribution of SNP effects&#39;)</code></pre>
<p><img src="figure/intro_gp.Rmd/unnamed-chunk-31-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Calculate the GEBVs using the SNP-effects (SNP-BLUPs)</p>
<pre class="r"><code># calculate the GEBVs using the SNP-effects (SNP-BLUPs)
g_rr&lt;-Z[names(g),]%*%u
all(rownames(g_rr)==names(g))</code></pre>
<pre><code>[1] TRUE</code></pre>
<pre class="r"><code>g_rr[1:5,]</code></pre>
<pre><code>        50395    95NA-00063   BAHKYEHEMAA BOUNOUA_BLANC       CH92108 
   -0.6734370     0.7155646     2.2350879     1.5461074     1.2510628 </code></pre>
<pre class="r"><code>g[1:5]</code></pre>
<pre><code>        50395    95NA-00063   BAHKYEHEMAA BOUNOUA_BLANC       CH92108 
   -0.6725989     0.7151010     2.2330024     1.5445541     1.2497041 </code></pre>
<pre class="r"><code># cor(g,g_rr) # [1,] 0.9999997
plot(x=g,y=g_rr,
     main = &#39;Proof that GEBV from GBLUP and RRBLUP are equal!&#39;,
     xlab=&quot;GEBV from GBLUP&quot;, ylab=&quot;GEBV from RRBLUP&quot;); 
abline(a=0,b=1,col=&#39;red&#39;)</code></pre>
<p><img src="figure/intro_gp.Rmd/unnamed-chunk-35-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="study-material" class="section level1">
<h1>Study material</h1>
<p>The Genomic Selection Manual, Chapter 10 has lot’s of “Recommended Literature” at the bottom of the section!</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.1 (2021-08-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] furrr_0.2.3              future_1.23.0            rsample_0.1.1           
 [4] sommer_4.1.5             crayon_1.4.2             lattice_0.20-45         
 [7] MASS_7.3-54              Matrix_1.4-0             genomicMateSelectR_0.2.0
[10] forcats_0.5.1            stringr_1.4.0            dplyr_1.0.7             
[13] purrr_0.3.4              readr_2.1.1              tidyr_1.1.4             
[16] tibble_3.1.6             ggplot2_3.3.5            tidyverse_1.3.1         
[19] workflowr_1.7.0         

loaded via a namespace (and not attached):
 [1] httr_1.4.2        sass_0.4.0        jsonlite_1.7.2    here_1.0.1       
 [5] modelr_0.1.8      bslib_0.3.1       assertthat_0.2.1  getPass_0.2-2    
 [9] highr_0.9         cellranger_1.1.0  yaml_2.2.1        globals_0.14.0   
[13] pillar_1.6.4      backports_1.4.1   glue_1.6.0        digest_0.6.29    
[17] promises_1.2.0.1  rvest_1.0.2       colorspace_2.0-2  htmltools_0.5.2  
[21] httpuv_1.6.5      pkgconfig_2.0.3   broom_0.7.11      listenv_0.8.0    
[25] haven_2.4.3       scales_1.1.1      processx_3.5.2    whisker_0.4      
[29] later_1.3.0       tzdb_0.2.0        git2r_0.29.0      generics_0.1.1   
[33] ellipsis_0.3.2    withr_2.4.3       cli_3.1.0         magrittr_2.0.1   
[37] readxl_1.3.1      evaluate_0.14     ps_1.6.0          parallelly_1.30.0
[41] fs_1.5.2          fansi_0.5.0       xml2_1.3.3        tools_4.1.1      
[45] hms_1.1.1         lifecycle_1.0.1   munsell_0.5.0     reprex_2.0.1     
[49] callr_3.7.0       compiler_4.1.1    jquerylib_0.1.4   rlang_0.4.12     
[53] grid_4.1.1        rstudioapi_0.13   rmarkdown_2.11    codetools_0.2-18 
[57] gtable_0.3.0      DBI_1.1.2         R6_2.5.1          lubridate_1.8.0  
[61] knitr_1.37        fastmap_1.1.0     utf8_1.2.2        rprojroot_2.0.2  
[65] stringi_1.7.6     parallel_4.1.1    Rcpp_1.0.7        vctrs_0.3.8      
[69] dbplyr_2.1.1      tidyselect_1.1.1  xfun_0.29        </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
