<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="wolfemd" />

<meta name="date" content="2022-05-09" />

<title>Parent-wise Cross-validation</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GSexample2022</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="download.html">Download</a>
</li>
<li>
  <a href="clean_phenos.html">Clean phenos</a>
</li>
<li>
  <a href="prep_genos.html">Prep genos</a>
</li>
<li>
  <a href="get_blups.html">Get BLUPs</a>
</li>
<li>
  <a href="intro_gp.html">Intro to GP</a>
</li>
<li>
  <a href="kfoldcrossval.html">K-fold Cross-val</a>
</li>
<li>
  <a href="get_gebv.html">Get GEBVs</a>
</li>
<li>
  <a href="parentwise_cv.html">Parent-wise CV</a>
</li>
<li>
  <a href="pred_crosses.html">Predict crosses</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Parent-wise Cross-validation</h1>
<h4 class="author">wolfemd</h4>
<h4 class="date">2022-05-09</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-05-09
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>GSexample2022/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220320code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20220320)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220320code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220320)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomwolfemdGSexample2022treebb2d98179db87acbe2189a400bac5d598f9119d8targetblankbb2d981a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/wolfemd/GSexample2022/tree/bb2d98179db87acbe2189a400bac5d598f9119d8" target="_blank">bb2d981</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomwolfemdGSexample2022treebb2d98179db87acbe2189a400bac5d598f9119d8targetblankbb2d981a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/wolfemd/GSexample2022/tree/bb2d98179db87acbe2189a400bac5d598f9119d8" target="_blank">bb2d981</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  .DS_Store
    Untracked:  Icon
    Untracked:  analysis/.DS_Store
    Untracked:  analysis/BreedBaseGenotypes_subset.log
    Untracked:  analysis/Icon
    Untracked:  analysis/images/Icon
    Untracked:  analysis/images/paste-2F34C15B.png
    Untracked:  analysis/images/paste-AD7AEEAC.png
    Untracked:  analysis/images/paste-CB5D91AC.png
    Untracked:  analysis/troubleshooting.Rmd
    Untracked:  code/Icon
    Untracked:  data/.DS_Store
    Untracked:  data/BreedBaseGenotypesDownload.positions
    Untracked:  data/BreedBaseGenotypesDownload.vcf
    Untracked:  data/BreedBaseGenotypes_subset.bed
    Untracked:  data/BreedBaseGenotypes_subset.bim
    Untracked:  data/BreedBaseGenotypes_subset.fam
    Untracked:  data/BreedBaseGenotypes_subset.hap.gz
    Untracked:  data/BreedBaseGenotypes_subset.log
    Untracked:  data/BreedBaseGenotypes_subset.nosex
    Untracked:  data/BreedBaseGenotypes_subset.positions
    Untracked:  data/BreedBaseGenotypes_subset.samples
    Untracked:  data/Icon
    Untracked:  data/metadata_cynthia.csv
    Untracked:  data/phenotype_cynthia.csv
    Untracked:  data/subset_unique_names_for_vcf.txt
    Untracked:  data/unique_names_for_vcf.txt
    Untracked:  data/vcf_colnames.txt
    Untracked:  out.log
    Untracked:  output/.DS_Store
    Untracked:  output/BreedBaseGenotypes_subset.genome
    Untracked:  output/BreedBaseGenotypes_subset.log
    Untracked:  output/BreedBaseGenotypes_subset.nosex
    Untracked:  output/Icon
    Untracked:  output/genomicPredictions.rds
    Untracked:  output/models_compared.Rdata
    Untracked:  output/parentWiseCV.rds
    Untracked:  output/standardCV.rds
    Untracked:  output/verified_ped.rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/parentwise_cv.Rmd</code>) and HTML (<code>docs/parentwise_cv.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/bb2d98179db87acbe2189a400bac5d598f9119d8/analysis/parentwise_cv.Rmd" target="_blank">bb2d981</a>
</td>
<td>
wolfemd
</td>
<td>
2022-05-09
</td>
<td>
Add the final two sections - parent-wise cross-validation plus predicting crosses
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="load-packages" class="section level1">
<h1>Load packages</h1>
<pre class="r"><code>suppressMessages(library(tidyverse)); 
suppressMessages(library(genomicMateSelectR)); </code></pre>
</div>
<div id="pedigree" class="section level1">
<h1>Pedigree</h1>
<p>Read pedigree</p>
<pre class="r"><code>ped&lt;-read_delim(here::here(&quot;data&quot;,&quot;pedigree.txt&quot;),delim = &quot;\t&quot;)</code></pre>
<pre><code>Rows: 1007 Columns: 4</code></pre>
<pre><code>── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;\t&quot;
chr (4): Accession, Female_Parent, Male_Parent, Cross_Type</code></pre>
<pre><code>
ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<p>Filter: Keep only complete pedigree records.</p>
<pre class="r"><code>ped %&lt;&gt;% 
     dplyr::select(-Cross_Type) %&gt;% 
     filter(!is.na(Female_Parent),
            !is.na(Male_Parent),
            Female_Parent!=&quot;?&quot;,
            Male_Parent!=&quot;?&quot;) %&gt;% 
     distinct</code></pre>
<pre class="r"><code>ped %&gt;% distinct(Female_Parent,Male_Parent) %&gt;% nrow()</code></pre>
<pre><code>[1] 478</code></pre>
<p>Summarize the pedigree</p>
<pre class="r"><code>ped %&gt;%   
     count(Female_Parent,Male_Parent) %&gt;% 
     arrange(desc(n))  %&gt;%  
     summary(.$n)</code></pre>
<pre><code> Female_Parent      Male_Parent              n        
 Length:478         Length:478         Min.   :1.000  
 Class :character   Class :character   1st Qu.:1.000  
 Mode  :character   Mode  :character   Median :1.000  
                                       Mean   :1.289  
                                       3rd Qu.:1.000  
                                       Max.   :7.000  </code></pre>
<p>Fully genotyped trios?</p>
<pre class="r"><code>dosages&lt;-readRDS(here::here(&quot;data&quot;,&quot;dosages.rds&quot;))
genotyped_gids&lt;-rownames(dosages)</code></pre>
<p>Are all of the entrees themselves genotyped?</p>
<pre class="r"><code>all(ped$Accession %in% genotyped_gids)</code></pre>
<pre><code>[1] TRUE</code></pre>
<pre class="r"><code>table(ped$Female_Parent %in% genotyped_gids)</code></pre>
<pre><code>
FALSE  TRUE 
  364   252 </code></pre>
<pre class="r"><code>table(ped$Male_Parent %in% genotyped_gids)</code></pre>
<pre><code>
FALSE  TRUE 
  337   279 </code></pre>
<pre class="r"><code>genotyped_ped&lt;-ped %&gt;% 
     filter(Accession %in% genotyped_gids,
            Female_Parent %in% genotyped_gids,
            Male_Parent %in% genotyped_gids)
genotyped_ped %&gt;% nrow()</code></pre>
<pre><code>[1] 155</code></pre>
<pre class="r"><code>genotyped_ped %&gt;% 
     count(Female_Parent,Male_Parent) %&gt;% 
     arrange(desc(n))  %&gt;%  
     summary(.$n)</code></pre>
<pre><code> Female_Parent      Male_Parent              n        
 Length:117         Length:117         Min.   :1.000  
 Class :character   Class :character   1st Qu.:1.000  
 Mode  :character   Mode  :character   Median :1.000  
                                       Mean   :1.325  
                                       3rd Qu.:1.000  
                                       Max.   :7.000  </code></pre>
<p>How many families have &gt;1 offspring?</p>
<pre class="r"><code>genotyped_ped %&gt;% 
     count(Female_Parent,Male_Parent) %&gt;% 
     filter(n&gt;1)</code></pre>
<pre><code># A tibble: 21 × 3
   Female_Parent        Male_Parent             n
   &lt;chr&gt;                &lt;chr&gt;               &lt;int&gt;
 1 IITA-TMS-IBA011371   IITA-TMS-IBA011371      4
 2 IITA-TMS-IBA020431   IITA-TMS-IBA030055A     2
 3 IITA-TMS-IBA030060   IITA-TMS-IBA010903      2
 4 IITA-TMS-IBA030060   IITA-TMS-IBA930265      2
 5 IITA-TMS-IBA030060   IITA-TMS-MM970043       3
 6 IITA-TMS-IBA30555    TMEB1                   2
 7 IITA-TMS-IBA4(2)1425 TMEB1                   3
 8 IITA-TMS-IBA8902195  IITA-TMS-IBA950379      2
 9 IITA-TMS-IBA9001554  IITA-TMS-IBA011659      2
10 IITA-TMS-IBA91934    TMEB1                   7
# … with 11 more rows</code></pre>
</div>
<div id="verify-pedigree-relationships" class="section level1">
<h1>Verify pedigree relationships</h1>
<p>My approach uses the <code>--genome</code> IBD calculator in the command-line program PLINK v1.9</p>
<p>Go over steps in manual: <a href="https://wolfemd.github.io/GenomicSelectionManual/parentwise_cross_val.html#verify-pedigree-relationships" class="uri">https://wolfemd.github.io/GenomicSelectionManual/parentwise_cross_val.html#verify-pedigree-relationships</a></p>
<div id="in-plink" class="section level2">
<h2>In plink</h2>
<p>In the terminal (not R) change directory</p>
<p>Go to the data/ directory where the VCF file is located</p>
<pre class="bash"><code>plink --vcf BreedBaseGenotypes_subset.vcf.gz \
     --make-bed --const-fid --keep-allele-order \
     --out BreedBaseGenotypes_subset</code></pre>
<p>still in the terminal</p>
<pre class="bash"><code>plink --bfile BreedBaseGenotypes_subset \
  --genome \
  --out ../output/BreedBaseGenotypes_subset;</code></pre>
</div>
<div id="verify-parent-offspring-relationships" class="section level2">
<h2>Verify parent-offspring relationships</h2>
<p>Back in R</p>
<pre class="r"><code>genome&lt;-read.table(here::here(&quot;output/&quot;,&quot;BreedBaseGenotypes_subset.genome&quot;),
                   stringsAsFactors = F,header = T) %&gt;% 
     as_tibble
genome %&gt;% head</code></pre>
<pre><code># A tibble: 6 × 14
   FID1 IID1     FID2 IID2    RT       EZ    Z0     Z1     Z2 PI_HAT   PHE   DST
  &lt;int&gt; &lt;chr&gt;   &lt;int&gt; &lt;chr&gt;   &lt;chr&gt; &lt;int&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt;
1     0 IITA-T…     0 IITA-T… OT        0 0.616 0.384  0       0.192    -1 0.727
2     0 IITA-T…     0 IITA-T… OT        0 0.663 0.337  0       0.168    -1 0.694
3     0 IITA-T…     0 IITA-T… OT        0 0.529 0.471  0       0.236    -1 0.769
4     0 IITA-T…     0 IITA-T… OT        0 0.607 0.283  0.110   0.252    -1 0.798
5     0 IITA-T…     0 IITA-T… OT        0 0.586 0.408  0.0054  0.210    -1 0.784
6     0 IITA-T…     0 IITA-T… OT        0 0     0.0613 0.939   0.969    -1 0.990
# … with 2 more variables: PPC &lt;dbl&gt;, RATIO &lt;dbl&gt;</code></pre>
<pre class="r"><code>dim(genome)</code></pre>
<pre><code>[1] 506521     14</code></pre>
<pre class="r"><code>ped %&gt;% 
     semi_join(genome %&gt;% rename(Accession=IID1,Female_Parent=IID2)) %&gt;% 
     left_join(genome %&gt;% rename(Accession=IID1,Female_Parent=IID2))</code></pre>
<pre><code>Joining, by = c(&quot;Accession&quot;, &quot;Female_Parent&quot;)
Joining, by = c(&quot;Accession&quot;, &quot;Female_Parent&quot;)</code></pre>
<pre><code># A tibble: 4 × 15
  Accession    Female_Parent  Male_Parent    FID1  FID2 RT       EZ     Z0    Z1
  &lt;chr&gt;        &lt;chr&gt;          &lt;chr&gt;         &lt;int&gt; &lt;int&gt; &lt;chr&gt; &lt;int&gt;  &lt;dbl&gt; &lt;dbl&gt;
1 IITA-TMS-BA… TMEB1          IITA-TMS-IBA…     0     0 OT        0 0.0248 0.927
2 IITA-TMS-IB… IITA-TMS-IBA9… IITA-TMS-IBA…     0     0 OT        0 0.685  0.315
3 IITA-TMS-IB… TMEB1          IITA-TMS-IBA…     0     0 OT        0 0.854  0.116
4 IITA-TMS-IB… IITA-TMS-IBA9… IITA-TMS-IBA…     0     0 OT        0 0      1    
# … with 6 more variables: Z2 &lt;dbl&gt;, PI_HAT &lt;dbl&gt;, PHE &lt;int&gt;, DST &lt;dbl&gt;,
#   PPC &lt;dbl&gt;, RATIO &lt;dbl&gt;</code></pre>
<pre class="r"><code># Confirm Female_Parent - Offspring Relationship
## In the plink genome file
## IID1 or IID2 could be the Accession or the Female_Parent
conf_female_ped&lt;-genotyped_ped %&gt;% 
     inner_join(genome %&gt;% 
                     rename(Accession=IID1,Female_Parent=IID2)) %&gt;% 
     bind_rows(genotyped_ped %&gt;% 
                    inner_join(genome %&gt;% 
                                    rename(Accession=IID2,
                                           Female_Parent=IID1))) %&gt;% 
     # Declare confirm-reject Accession-Female_Parent
     mutate(ConfirmFemaleParent=case_when(Z0&lt;0.32 &amp; Z1&gt;0.67~&quot;Confirm&quot;, 
   # Relatedness coeff differ if the Accession is the result of a self-cross
          Male_Parent==Female_Parent &amp; PI_HAT&gt;0.6 &amp; Z0&lt;0.3 &amp; Z2&gt;0.32~&quot;Confirm&quot;,
   TRUE~&quot;Reject&quot;)) %&gt;% 
     dplyr::select(Accession,Female_Parent,ConfirmFemaleParent)</code></pre>
<pre><code>Joining, by = c(&quot;Accession&quot;, &quot;Female_Parent&quot;)
Joining, by = c(&quot;Accession&quot;, &quot;Female_Parent&quot;)</code></pre>
<pre class="r"><code>## Now do the same for the Accession-Male_Parent relationships
conf_male_ped&lt;-genotyped_ped %&gt;% 
     inner_join(genome %&gt;% 
                     rename(Accession=IID1,Male_Parent=IID2)) %&gt;% 
     bind_rows(genotyped_ped %&gt;% 
                    inner_join(genome %&gt;% 
                                    rename(Accession=IID2,Male_Parent=IID1))) %&gt;% 
     # Declare confirm-reject Accession-Female_Parent
     mutate(ConfirmMaleParent=case_when(Z0&lt;0.32 &amp; Z1&gt;0.67~&quot;Confirm&quot;, 
     # Relatedness coeff differ if the Accession is the result of a self-cross
          Male_Parent==Female_Parent &amp; PI_HAT&gt;0.6 &amp; Z0&lt;0.3 &amp; Z2&gt;0.32~&quot;Confirm&quot;,
                                          TRUE~&quot;Reject&quot;)) %&gt;% 
     dplyr::select(Accession,Male_Parent,ConfirmMaleParent)</code></pre>
<pre><code>Joining, by = c(&quot;Accession&quot;, &quot;Male_Parent&quot;)
Joining, by = c(&quot;Accession&quot;, &quot;Male_Parent&quot;)</code></pre>
<pre class="r"><code># Now join the confirmed female and male relationships
# This regenerates the original &quot;genotyped_ped&quot; with two added columns
confirmed_ped&lt;-conf_female_ped %&gt;% 
     left_join(conf_male_ped) %&gt;% 
     relocate(Male_Parent,.before = &quot;ConfirmFemaleParent&quot;)</code></pre>
<pre><code>Joining, by = &quot;Accession&quot;</code></pre>
<p>So, how well supported are the pedigree relationships according to my approach?</p>
<pre class="r"><code>confirmed_ped %&gt;% 
     count(ConfirmFemaleParent,ConfirmMaleParent) %&gt;% 
     mutate(Prop=round(n/sum(n),2))</code></pre>
<pre><code># A tibble: 4 × 4
  ConfirmFemaleParent ConfirmMaleParent     n  Prop
  &lt;chr&gt;               &lt;chr&gt;             &lt;int&gt; &lt;dbl&gt;
1 Confirm             Confirm             118  0.76
2 Confirm             Reject               11  0.07
3 Reject              Confirm               8  0.05
4 Reject              Reject               18  0.12</code></pre>
</div>
<div id="subset-to-fully-validated-trios" class="section level2">
<h2>Subset to fully-validated trios</h2>
<p>Remove any without both parents confirmed.</p>
<pre class="r"><code>valid_ped&lt;-confirmed_ped %&gt;% 
     filter(ConfirmFemaleParent==&quot;Confirm&quot;,
         ConfirmMaleParent==&quot;Confirm&quot;) %&gt;% 
     dplyr::select(-contains(&quot;Confirm&quot;))</code></pre>
<pre class="r"><code>valid_ped %&gt;% nrow()</code></pre>
<pre><code>[1] 118</code></pre>
<pre class="r"><code>valid_ped %&gt;% 
     count(Female_Parent,Male_Parent) %&gt;% 
     filter(n&gt;1)</code></pre>
<pre><code># A tibble: 19 × 3
   Female_Parent        Male_Parent             n
   &lt;chr&gt;                &lt;chr&gt;               &lt;int&gt;
 1 IITA-TMS-IBA011371   IITA-TMS-IBA011371      4
 2 IITA-TMS-IBA020431   IITA-TMS-IBA030055A     2
 3 IITA-TMS-IBA030060   IITA-TMS-IBA010903      2
 4 IITA-TMS-IBA030060   IITA-TMS-IBA930265      2
 5 IITA-TMS-IBA030060   IITA-TMS-MM970043       3
 6 IITA-TMS-IBA4(2)1425 TMEB1                   2
 7 IITA-TMS-IBA8902195  IITA-TMS-IBA950379      2
 8 IITA-TMS-IBA9001554  IITA-TMS-IBA011659      2
 9 IITA-TMS-IBA91934    TMEB1                   6
10 IITA-TMS-IBA940561   IITA-TMS-IBA940263      2
11 IITA-TMS-IBA950971   IITA-TMS-IBA940561      3
12 IITA-TMS-IBA961089A  IITA-TMS-IBA961089A     2
13 IITA-TMS-IBA961632   IITA-TMS-IBA000070      2
14 IITA-TMS-IBA961632   IITA-TMS-IBA030055A     3
15 IITA-TMS-IBA972205   TMEB1937                5
16 IITA-TMS-ZAR930151   IITA-TMS-MM970043       2
17 TMS13F1307P0004      TMS13F1343P0044         2
18 TMS14F1016P0006      TMS14F1035P0004         2
19 TMS14F1255P0005      TMS13F1343P0044         2</code></pre>
<pre class="r"><code>valid_ped %&gt;% 
     count(Female_Parent,Male_Parent) %&gt;% 
     filter(n&gt;2)</code></pre>
<pre><code># A tibble: 6 × 3
  Female_Parent      Male_Parent             n
  &lt;chr&gt;              &lt;chr&gt;               &lt;int&gt;
1 IITA-TMS-IBA011371 IITA-TMS-IBA011371      4
2 IITA-TMS-IBA030060 IITA-TMS-MM970043       3
3 IITA-TMS-IBA91934  TMEB1                   6
4 IITA-TMS-IBA950971 IITA-TMS-IBA940561      3
5 IITA-TMS-IBA961632 IITA-TMS-IBA030055A     3
6 IITA-TMS-IBA972205 TMEB1937                5</code></pre>
</div>
<div id="write-validated-pedigree" class="section level2">
<h2>Write validated pedigree</h2>
<pre class="r"><code>saveRDS(valid_ped,here::here(&quot;output&quot;,&quot;verified_ped.rds&quot;))</code></pre>
</div>
</div>
<div id="parent-wise-cross-validation" class="section level1">
<h1>Parent-wise cross-validation</h1>
<div id="set-up" class="section level2">
<h2>Set-up</h2>
<pre class="r"><code># Load verified ped
ped&lt;-readRDS(here::here(&quot;output&quot;,&quot;verified_ped.rds&quot;)) %&gt;% 
     # Rename things to match genomicMateSelectR::runParentWiseCrossVal()
     rename(GID=Accession,
            sireID=Male_Parent,
            damID=Female_Parent)
# Keep only families with _at least_ 2 offspring
ped %&lt;&gt;% 
     semi_join(ped %&gt;% count(sireID,damID) %&gt;% filter(n&gt;1) %&gt;% ungroup())</code></pre>
<pre><code>Joining, by = c(&quot;damID&quot;, &quot;sireID&quot;)</code></pre>
<pre class="r"><code># GENOMIC RELATIONSHIP MATRIX
grms&lt;-list(A=readRDS(file=here::here(&quot;output&quot;,&quot;kinship_add.rds&quot;)))

# BLUPs
blups&lt;-readRDS(here::here(&quot;output&quot;,&quot;blups.rds&quot;)) %&gt;% 
     # based on cross-validation, decided to exclude MCMDS from this analysis
     filter(Trait != &quot;MCMDS&quot;) %&gt;% 
     # need to rename the &quot;blups&quot; list to comply with the runCrossVal function
     rename(TrainingData=blups) %&gt;% 
     dplyr::select(Trait,TrainingData) %&gt;% 
     # need also to remove phenotyped-but-not-genotyped lines
     mutate(TrainingData=map(TrainingData,
                             ~filter(.,germplasmName %in% rownames(grms$A)) %&gt;% 
                                  # rename the germplasmName column to GID
                                  rename(GID=germplasmName))) %&gt;% 
     # It seems actually that runParentWiseCrossVal() wnats this column named &quot;blups&quot;
     rename(blups=TrainingData)

# DOSAGE MATRIX
## Dosages are also needed inside the runParentWiseCrossVal() function
## Reason is that they are used to extra SNP effects from GBLUP models
dosages&lt;-readRDS(here::here(&quot;data&quot;,&quot;dosages.rds&quot;))

# HAPLOTYPE MATRIX
## keep only haplos for parents-in-the-pedigree
## those which will be used in prediction, saves memory
haploMat&lt;-readRDS(file=here::here(&quot;data&quot;,&quot;haplotypes.rds&quot;))
parents&lt;-union(ped$sireID,ped$damID) 
parenthaps&lt;-sort(c(paste0(parents,&quot;_HapA&quot;),
                   paste0(parents,&quot;_HapB&quot;)))
haploMat&lt;-haploMat[parenthaps,]

# SELECTION INDEX
SIwts&lt;-c(DM=15,
         logFYLD=20,
         logDYLD=20)</code></pre>
<pre class="r"><code># RECOMBINATION FREQUENCY MATRIX
recombFreqMat&lt;-readRDS(file=here::here(&quot;output&quot;,&quot;recombFreqMat_1minus2c.rds&quot;))</code></pre>
<pre class="r"><code>starttime&lt;-proc.time()[3]
parentWiseCV&lt;-runParentWiseCrossVal(nrepeats=2,nfolds=5,seed=121212,
                                    modelType=&quot;A&quot;,
                                    ncores=10,
                                    ped=ped,
                                    blups=blups,
                                    dosages=dosages,
                                    haploMat=haploMat,
                                    grms=grms,
                                    recombFreqMat = recombFreqMat,
                                    selInd = TRUE, SIwts = SIwts)</code></pre>
<pre><code>Loading required package: rsample</code></pre>
<pre><code>[1] &quot;Set-up parent-wise cross-validation folds&quot;
[1] &quot;Fitting models to get marker effects&quot;</code></pre>
<pre><code>Joining, by = &quot;Trait&quot;</code></pre>
<pre><code>Loading required package: furrr</code></pre>
<pre><code>Loading required package: future</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -3.71141   20:22:44      0           0
    2      -3.70229   20:22:44      0           0
    3      -3.69222   20:22:44      0           0
    4      -3.68241   20:22:44      0           0
    5      -3.67488   20:22:44      0           0
    6      -3.66862   20:22:44      0           0
    7      -3.66307   20:22:44      0           0
    8      -3.65783   20:22:44      0           0
    9      -3.65261   20:22:44      0           1
    10      -3.64828   20:22:44      0           1
    11      -3.64828   20:22:44      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -2.24966   20:22:59      0           0
    2      -2.20818   20:22:59      0           0
    3      -2.14978   20:22:59      0           1
    4      -2.13835   20:22:59      0           1
    5      -2.13835   20:22:59      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -1.80555   20:23:20      0           0
    2      -1.78463   20:23:20      0           0
    3      -1.76522   20:23:20      0           0
    4      -1.75075   20:23:20      0           1
    5      -1.74321   20:23:20      0           1
    6      -1.74321   20:23:20      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -17.426   20:23:41      0           0
    2      -17.4165   20:23:41      0           0
    3      -17.412   20:23:41      0           0
    4      -17.411   20:23:41      0           0
    5      -17.4109   20:23:41      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -5.39846   20:24:1      0           0
    2      -5.18186   20:24:1      0           1
    3      -5.07868   20:24:1      0           1
    4      -5.07868   20:24:1      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -4.16748   20:24:20      0           0
    2      -4.13967   20:24:20      0           0
    3      -4.12459   20:24:20      0           0
    4      -4.11824   20:24:20      0           0
    5      -4.11624   20:24:20      0           0
    6      -4.11548   20:24:20      0           0</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre><code>
Attaching package: ‘Matrix’</code></pre>
<pre><code>The following objects are masked from ‘package:tidyr’:

    expand, pack, unpack</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: lattice</code></pre>
<pre><code>Loading required package: crayon</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -2.43093   20:22:45      0           0
    2      -2.26507   20:22:45      0           0
    3      -2.19404   20:22:45      0           1
    4      -2.18724   20:22:45      0           1
    5      -2.18724   20:22:45      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -3.34368   20:23:1      0           0
    2      -3.32559   20:23:1      0           0
    3      -3.31357   20:23:1      0           0
    4      -3.3066   20:23:1      0           0
    5      -3.30323   20:23:1      0           0
    6      -3.30125   20:23:1      0           0
    7      -3.29994   20:23:1      0           0
    8      -3.29899   20:23:1      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -1.97673   20:23:22      0           0
    2      -1.96142   20:23:22      0           0
    3      -1.95634   20:23:22      0           0
    4      -1.95547   20:23:22      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -19.4408   20:23:43      0           0
    2      -19.151   20:23:43      0           0
    3      -18.9227   20:23:43      0           0
    4      -18.7834   20:23:43      0           0
    5      -18.7251   20:23:43      0           0
    6      -18.7001   20:23:43      0           0
    7      -18.6894   20:23:43      0           0
    8      -18.6848   20:23:43      0           0
    9      -18.6828   20:23:43      0           0
    10      -18.682   20:23:43      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -3.98576   20:24:3      0           0
    2      -3.9837   20:24:3      0           0
    3      -3.98346   20:24:3      0           0
    4      -3.98346   20:24:3      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -3.57202   20:24:21      0           0
    2      -3.5048   20:24:21      0           0
    3      -3.49365   20:24:21      0           0
    4      -3.49345   20:24:21      0           0</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre><code>
Attaching package: ‘Matrix’</code></pre>
<pre><code>The following objects are masked from ‘package:tidyr’:

    expand, pack, unpack</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: lattice</code></pre>
<pre><code>Loading required package: crayon</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -2.61322   20:22:46      0           1
    2      -2.1819   20:22:46      0           1
    3      -2.1819   20:22:46      0           1
    4      -2.1819   20:22:46      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -1.33319   20:23:2      0           1
    2      -1.23189   20:23:2      0           1
    3      -1.23189   20:23:2      0           1
    4      -1.23189   20:23:2      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -1.39708   20:23:23      0           1
    2      -1.26254   20:23:23      0           1
    3      -1.26254   20:23:23      0           1
    4      -1.26254   20:23:23      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -17.0073   20:23:44      0           0
    2      -16.9501   20:23:44      0           0
    3      -16.8925   20:23:44      0           0
    4      -16.847   20:23:44      0           0
    5      -16.823   20:23:44      0           0
    6      -16.8108   20:23:44      0           0
    7      -16.8049   20:23:44      0           0
    8      -16.802   20:23:44      0           0
    9      -16.8006   20:23:44      0           0
    10      -16.8   20:23:44      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -4.82084   20:24:4      0           0
    2      -4.46573   20:24:4      0           1
    3      -4.38128   20:24:4      0           1
    4      -4.38128   20:24:4      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -3.6018   20:24:22      0           0
    2      -3.43307   20:24:22      0           0
    3      -3.37182   20:24:22      0           0
    4      -3.36592   20:24:22      0           0
    5      -3.36592   20:24:22      0           0</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre><code>
Attaching package: ‘Matrix’</code></pre>
<pre><code>The following objects are masked from ‘package:tidyr’:

    expand, pack, unpack</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: lattice</code></pre>
<pre><code>Loading required package: crayon</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -4.62508   20:22:46      0           0
    2      -4.59644   20:22:46      0           0
    3      -4.59392   20:22:46      0           0
    4      -4.59392   20:22:46      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -2.80799   20:23:3      0           0
    2      -2.68322   20:23:3      0           0
    3      -2.5353   20:23:3      0           1
    4      -2.53098   20:23:3      0           1
    5      -2.53098   20:23:3      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -3.022   20:23:24      0           0
    2      -3.00287   20:23:24      0           0
    3      -2.97806   20:23:24      0           0
    4      -2.94237   20:23:24      0           1
    5      -2.90687   20:23:24      0           1
    6      -2.90687   20:23:24      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -15.8294   20:23:46      0           0
    2      -15.7112   20:23:46      0           0
    3      -15.6465   20:23:46      0           0
    4      -15.6238   20:23:46      0           0
    5      -15.6192   20:23:46      0           0
    6      -15.6183   20:23:46      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -4.93524   20:24:5      0           1
    2      -3.78587   20:24:5      0           1
    3      -3.78587   20:24:5      0           1
    4      -3.78587   20:24:5      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -3.40761   20:24:24      0           1
    2      -2.39443   20:24:24      0           1
    3      -2.39443   20:24:24      0           1
    4      -2.39443   20:24:24      0           1</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre><code>
Attaching package: ‘Matrix’</code></pre>
<pre><code>The following objects are masked from ‘package:tidyr’:

    expand, pack, unpack</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: lattice</code></pre>
<pre><code>Loading required package: crayon</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -5.87152   20:22:47      0           0
    2      -5.86814   20:22:47      0           0
    3      -5.8653   20:22:47      0           0
    4      -5.86356   20:22:47      0           0
    5      -5.8629   20:22:47      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -0.5   20:23:4      0           0
    2      -0.5   20:23:4      0           0
    3      -0.5   20:23:4      0           0
    4      -0.5   20:23:4      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -0.5   20:23:25      0           0
    2      -0.5   20:23:25      0           0
    3      -0.5   20:23:25      0           0
    4      -0.5   20:23:25      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -13.8957   20:23:47      0           0
    2      -13.8667   20:23:47      0           0
    3      -13.8406   20:23:47      0           0
    4      -13.8217   20:23:47      0           0
    5      -13.8119   20:23:47      0           0
    6      -13.8064   20:23:47      0           0
    7      -13.8032   20:23:47      0           0
    8      -13.8013   20:23:47      0           0
    9      -13.8001   20:23:47      0           0
    10      -13.7994   20:23:47      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -7.05175   20:24:6      0           0
    2      -6.53049   20:24:6      0           1
    3      -6.45278   20:24:6      0           1
    4      -6.45278   20:24:6      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -4.87573   20:24:25      0           0
    2      -4.49937   20:24:25      0           0
    3      -4.39236   20:24:25      0           1
    4      -4.39169   20:24:25      0           1</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre><code>
Attaching package: ‘Matrix’</code></pre>
<pre><code>The following objects are masked from ‘package:tidyr’:

    expand, pack, unpack</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: lattice</code></pre>
<pre><code>Loading required package: crayon</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -5.86527   20:22:48      0           0
    2      -5.646   20:22:48      0           0
    3      -5.51752   20:22:48      0           0
    4      -5.48079   20:22:48      0           0
    5      -5.47678   20:22:48      0           0
    6      -5.47635   20:22:48      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -2.00835   20:23:5      0           0
    2      -1.99181   20:23:5      0           1
    3      -1.97135   20:23:5      0           1
    4      -1.97135   20:23:5      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -1.15535   20:23:27      0           0
    2      -1.14758   20:23:27      0           0
    3      -1.14217   20:23:27      0           0
    4      -1.13978   20:23:27      0           0
    5      -1.13923   20:23:27      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -14.5687   20:23:48      0           0
    2      -14.4816   20:23:48      0           0
    3      -14.4286   20:23:48      0           0
    4      -14.4092   20:23:48      0           0
    5      -14.4056   20:23:48      0           0
    6      -14.4049   20:23:48      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -5.61355   20:24:8      1           0
    2      -5.10668   20:24:8      1           1
    3      -5.07505   20:24:8      1           1
    4      -5.07505   20:24:8      1           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -4.87269   20:24:26      0           0
    2      -4.56413   20:24:26      0           0
    3      -4.48859   20:24:26      0           1
    4      -4.48815   20:24:26      0           1</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre><code>
Attaching package: ‘Matrix’</code></pre>
<pre><code>The following objects are masked from ‘package:tidyr’:

    expand, pack, unpack</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: lattice</code></pre>
<pre><code>Loading required package: crayon</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -1.95471   20:22:49      0           0
    2      -1.94916   20:22:49      0           0
    3      -1.94734   20:22:49      0           0
    4      -1.94709   20:22:49      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -3.12389   20:23:6      0           0
    2      -3.11939   20:23:6      0           0
    3      -3.11098   20:23:6      0           0
    4      -3.09565   20:23:6      0           1
    5      -3.08166   20:23:6      0           1
    6      -3.08166   20:23:6      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -2.6904   20:23:29      0           0
    2      -2.67494   20:23:29      0           0
    3      -2.64801   20:23:29      0           1
    4      -2.61336   20:23:29      0           1
    5      -2.61336   20:23:29      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -18.9371   20:23:49      0           0
    2      -18.7405   20:23:49      0           0
    3      -18.6451   20:23:49      0           0
    4      -18.6163   20:23:49      0           0
    5      -18.6118   20:23:49      0           0
    6      -18.6111   20:23:49      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -3.77109   20:24:9      0           1
    2      -3.07247   20:24:9      0           1
    3      -3.07247   20:24:9      0           1
    4      -3.07247   20:24:9      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -2.52503   20:24:27      0           0
    2      -1.98141   20:24:27      0           1
    3      -1.9165   20:24:27      0           1
    4      -1.9165   20:24:27      0           1</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre><code>
Attaching package: ‘Matrix’</code></pre>
<pre><code>The following objects are masked from ‘package:tidyr’:

    expand, pack, unpack</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: lattice</code></pre>
<pre><code>Loading required package: crayon</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -5.89645   20:22:50      0           0
    2      -5.87231   20:22:50      0           0
    3      -5.86687   20:22:50      0           0
    4      -5.86665   20:22:50      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -3.6093   20:23:7      0           0
    2      -3.05759   20:23:7      0           0
    3      -1.99963   20:23:7      0           1
    4      -1.59665   20:23:7      0           1
    5      -1.59665   20:23:7      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -3.30312   20:23:30      0           0
    2      -3.03397   20:23:30      0           0
    3      -2.60515   20:23:30      0           0
    4      -1.79851   20:23:30      0           1
    5      -1.486   20:23:30      0           1
    6      -1.486   20:23:30      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -15.1543   20:23:50      0           0
    2      -15.0122   20:23:50      0           0
    3      -14.904   20:23:50      0           0
    4      -14.8436   20:23:50      0           0
    5      -14.8218   20:23:50      0           0
    6      -14.8136   20:23:50      0           0
    7      -14.8106   20:23:50      0           0
    8      -14.8094   20:23:50      0           0
    9      -14.8089   20:23:50      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -4.75855   20:24:10      0           0
    2      -4.38903   20:24:10      0           1
    3      -4.30154   20:24:10      0           1
    4      -4.30154   20:24:10      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -4.13774   20:24:28      0           0
    2      -4.03445   20:24:28      0           0
    3      -3.97666   20:24:28      0           0
    4      -3.96326   20:24:28      0           0
    5      -3.96276   20:24:28      0           0</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre><code>
Attaching package: ‘Matrix’</code></pre>
<pre><code>The following objects are masked from ‘package:tidyr’:

    expand, pack, unpack</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: lattice</code></pre>
<pre><code>Loading required package: crayon</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -4.507   20:22:50      0           0
    2      -4.49123   20:22:50      0           0
    3      -4.48358   20:22:50      0           0
    4      -4.48152   20:22:50      0           0
    5      -4.48126   20:22:50      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -1.77781   20:23:8      0           1
    2      -1.65331   20:23:8      0           1
    3      -1.65331   20:23:8      0           1
    4      -1.65331   20:23:8      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -2.02129   20:23:32      0           0
    2      -2.01602   20:23:32      0           0
    3      -2.01082   20:23:32      0           0
    4      -2.00674   20:23:32      0           0
    5      -2.00453   20:23:32      0           0
    6      -2.00332   20:23:32      0           0
    7      -2.00265   20:23:32      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -15.7852   20:23:51      0           0
    2      -15.5655   20:23:51      0           0
    3      -15.3735   20:23:51      0           0
    4      -15.2588   20:23:51      0           0
    5      -15.2193   20:23:51      0           0
    6      -15.2066   20:23:51      0           0
    7      -15.2027   20:23:51      0           0
    8      -15.2016   20:23:51      0           0
    9      -15.2012   20:23:51      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -5.07884   20:24:11      0           0
    2      -4.79738   20:24:11      0           1
    3      -4.73054   20:24:11      0           1
    4      -4.73054   20:24:11      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -3.18385   20:24:30      0           0
    2      -2.71601   20:24:30      0           1
    3      -2.65366   20:24:30      0           1
    4      -2.65366   20:24:30      0           1</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre><code>
Attaching package: ‘Matrix’</code></pre>
<pre><code>The following objects are masked from ‘package:tidyr’:

    expand, pack, unpack</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: lattice</code></pre>
<pre><code>Loading required package: crayon</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -3.38334   20:22:51      0           0
    2      -3.20766   20:22:51      0           1
    3      -3.1676   20:22:51      0           1
    4      -3.1676   20:22:51      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -2.89715   20:23:9      0           0
    2      -2.72503   20:23:9      0           0
    3      -2.51412   20:23:9      0           0
    4      -2.16229   20:23:9      0           1
    5      -1.84532   20:23:9      0           1
    6      -1.84532   20:23:9      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -2.66857   20:23:32      0           0
    2      -2.65389   20:23:32      0           0
    3      -2.63803   20:23:32      0           0
    4      -2.62064   20:23:32      0           0
    5      -2.60221   20:23:32      0           0
    6      -2.57762   20:23:32      0           0
    7      -2.53725   20:23:32      0           0
    8      -2.45651   20:23:32      0           0
    9      -2.26384   20:23:32      0           1
    10      -1.89825   20:23:32      0           1
    11      -1.89825   20:23:32      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -16.3768   20:23:52      0           0
    2      -16.371   20:23:52      0           0
    3      -16.3665   20:23:52      0           0
    4      -16.364   20:23:52      0           0
    5      -16.3631   20:23:52      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -7.14619   20:24:12      0           0
    2      -6.76967   20:24:12      0           1
    3      -6.73009   20:24:12      0           1
    4      -6.73009   20:24:12      0           1
iteration    LogLik     wall    cpu(sec)   restrained
    1      -6.1113   20:24:31      0           0
    2      -5.95766   20:24:31      0           0
    3      -5.91088   20:24:31      0           0
    4      -5.90926   20:24:31      0           0
    5      -5.90923   20:24:31      0           0</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: Matrix</code></pre>
<pre><code>
Attaching package: ‘Matrix’</code></pre>
<pre><code>The following objects are masked from ‘package:tidyr’:

    expand, pack, unpack</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: lattice</code></pre>
<pre><code>Loading required package: crayon</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>Loading required package: sommer</code></pre>
<pre><code>Loading required package: MASS</code></pre>
<pre><code>
Attaching package: ‘MASS’</code></pre>
<pre><code>The following object is masked from ‘package:dplyr’:

    select</code></pre>
<pre><code>[1] &quot;Marker-effects Computed. Took  0.0343 hrs&quot;
[1] &quot;Predicting cross variances and covariances&quot;</code></pre>
<pre><code>Joining, by = c(&quot;Repeat&quot;, &quot;Fold&quot;)</code></pre>
<pre><code>[1] &quot;Done predicting fam vars. Took 0.11 mins for 7 crosses&quot;
[1] &quot;Done predicting fam vars. Took 0.13 mins for 8 crosses&quot;
[1] &quot;Done predicting fam vars. Took 0.11 mins for 7 crosses&quot;
[1] &quot;Done predicting fam vars. Took 0.1 mins for 6 crosses&quot;
[1] &quot;Done predicting fam vars. Took 0.09 mins for 5 crosses&quot;
[1] &quot;Done predicting fam vars. Took 0.13 mins for 8 crosses&quot;
[1] &quot;Done predicting fam vars. Took 0.1 mins for 6 crosses&quot;
[1] &quot;Done predicting fam vars. Took 0.13 mins for 8 crosses&quot;
[1] &quot;Done predicting fam vars. Took 0.11 mins for 7 crosses&quot;
[1] &quot;Done predicting fam vars. Took 0.09 mins for 5 crosses&quot;
[1] &quot;Cross variance parameters predicted. Took  0.01845 hrs&quot;
[1] &quot;Predicting cross means&quot;</code></pre>
<pre><code>Joining, by = c(&quot;Repeat&quot;, &quot;Fold&quot;)</code></pre>
<pre><code>[1] &quot;Cross means predicted. Took  0.0062 hrs&quot;
[1] &quot;Compute prediction accuracies and wrap up.&quot;</code></pre>
<pre><code>Joining, by = c(&quot;damID&quot;, &quot;sireID&quot;)</code></pre>
<pre><code>Joining, by = c(&quot;Repeat&quot;, &quot;Fold&quot;, &quot;modelType&quot;)</code></pre>
<pre><code>Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;</code></pre>
<pre><code>Joining, by = c(&quot;Repeat&quot;, &quot;Fold&quot;, &quot;modelType&quot;, &quot;predOf&quot;, &quot;sireID&quot;, &quot;damID&quot;, &quot;Trait1&quot;, &quot;Trait2&quot;)</code></pre>
<pre><code>Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful</code></pre>
<pre><code>Joining, by = c(&quot;damID&quot;, &quot;sireID&quot;)</code></pre>
<pre><code>Joining, by = c(&quot;Repeat&quot;, &quot;Fold&quot;, &quot;modelType&quot;)</code></pre>
<pre><code>Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;
Joining, by = &quot;GID&quot;</code></pre>
<pre><code>Joining, by = c(&quot;Repeat&quot;, &quot;Fold&quot;, &quot;modelType&quot;, &quot;predOf&quot;, &quot;sireID&quot;, &quot;damID&quot;, &quot;Trait&quot;)</code></pre>
<pre><code>Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful

Warning in cov2cor(cov): diag(.) had 0 or NA entries; non-finite result is
doubtful</code></pre>
<pre><code>[1] &quot;Accuracies predicted. Took  0.06084 hrs total.Goodbye!&quot;</code></pre>
<pre class="r"><code>elapsed&lt;-proc.time()[3]-starttime; elapsed/60</code></pre>
<pre><code> elapsed 
3.650183 </code></pre>
<pre class="r"><code>saveRDS(parentWiseCV,file = here::here(&quot;output&quot;,&quot;parentWiseCV.rds&quot;))</code></pre>
<pre class="r"><code>parentWiseCV$meanPredAccuracy %&gt;% head</code></pre>
<pre><code># A tibble: 6 × 7
  Repeat  Fold  modelType predOf Trait   predVSobs        AccuracyEst
  &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;   &lt;list&gt;                 &lt;dbl&gt;
1 Repeat1 Fold1 A         MeanBV SELIND  &lt;tibble [7 × 5]&gt;     NaN    
2 Repeat1 Fold1 A         MeanBV DM      &lt;tibble [7 × 5]&gt;     NaN    
3 Repeat1 Fold1 A         MeanBV logDYLD &lt;tibble [7 × 5]&gt;     NaN    
4 Repeat1 Fold1 A         MeanBV logFYLD &lt;tibble [7 × 5]&gt;     NaN    
5 Repeat1 Fold2 A         MeanBV SELIND  &lt;tibble [8 × 5]&gt;      -0.018
6 Repeat1 Fold2 A         MeanBV DM      &lt;tibble [8 × 5]&gt;     NaN    </code></pre>
<pre class="r"><code>parentWiseCV$varPredAccuracy %&gt;% head</code></pre>
<pre><code># A tibble: 6 × 8
  Repeat  Fold  modelType predOf Trait1  Trait2  predVSobs        AccuracyEst
  &lt;chr&gt;   &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;   &lt;list&gt;                 &lt;dbl&gt;
1 Repeat1 Fold1 A         VarBV  SELIND  SELIND  &lt;tibble [7 × 5]&gt;         NaN
2 Repeat1 Fold1 A         VarBV  DM      DM      &lt;tibble [7 × 5]&gt;         NaN
3 Repeat1 Fold1 A         VarBV  DM      logDYLD &lt;tibble [7 × 5]&gt;         NaN
4 Repeat1 Fold1 A         VarBV  DM      logFYLD &lt;tibble [7 × 5]&gt;         NaN
5 Repeat1 Fold1 A         VarBV  logDYLD logDYLD &lt;tibble [7 × 5]&gt;         NaN
6 Repeat1 Fold1 A         VarBV  logDYLD logFYLD &lt;tibble [7 × 5]&gt;         NaN</code></pre>
<pre class="r"><code>parentWiseCV$meanPredAccuracy %&gt;% 
     ggplot(.,aes(x=Trait,y=AccuracyEst,fill=Trait)) + geom_boxplot() + 
     labs(title=&quot;Accuracy Predicting Family Means&quot;)</code></pre>
<pre><code>Warning: Removed 22 rows containing non-finite values (stat_boxplot).</code></pre>
<p><img src="figure/parentwise_cv.Rmd/unnamed-chunk-32-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>parentWiseCV$varPredAccuracy %&gt;% 
     # this will format the two column information 
     # indicating variances and covariances
     # into a single variable for the plot
     mutate(VarParam=paste0(Trait1,&quot;\n&quot;,Trait2)) %&gt;% 
     ggplot(.,aes(x=VarParam,y=AccuracyEst,fill=VarParam)) + geom_boxplot()</code></pre>
<pre><code>Warning: Removed 50 rows containing non-finite values (stat_boxplot).</code></pre>
<p><img src="figure/parentwise_cv.Rmd/unnamed-chunk-33-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.1 (2021-08-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] furrr_0.2.3              future_1.23.0            rsample_0.1.1           
 [4] genomicMateSelectR_0.2.0 forcats_0.5.1            stringr_1.4.0           
 [7] dplyr_1.0.7              purrr_0.3.4              readr_2.1.1             
[10] tidyr_1.1.4              tibble_3.1.6             ggplot2_3.3.5           
[13] tidyverse_1.3.1          workflowr_1.7.0         

loaded via a namespace (and not attached):
 [1] nlme_3.1-153      fs_1.5.2          lubridate_1.8.0   bit64_4.0.5      
 [5] httr_1.4.2        rprojroot_2.0.2   tools_4.1.1       backports_1.4.1  
 [9] bslib_0.3.1       utf8_1.2.2        R6_2.5.1          DBI_1.1.2        
[13] colorspace_2.0-2  withr_2.4.3       tidyselect_1.1.1  mnormt_2.0.2     
[17] processx_3.5.2    bit_4.0.4         compiler_4.1.1    git2r_0.29.0     
[21] cli_3.1.0         rvest_1.0.2       xml2_1.3.3        labeling_0.4.2   
[25] sass_0.4.0        scales_1.1.1      psych_2.1.9       callr_3.7.0      
[29] digest_0.6.29     rmarkdown_2.11    pkgconfig_2.0.3   htmltools_0.5.2  
[33] parallelly_1.30.0 highr_0.9         dbplyr_2.1.1      fastmap_1.1.0    
[37] rlang_0.4.12      readxl_1.3.1      rstudioapi_0.13   farver_2.1.0     
[41] jquerylib_0.1.4   generics_0.1.1    jsonlite_1.7.2    vroom_1.5.7      
[45] magrittr_2.0.1    Rcpp_1.0.7        munsell_0.5.0     fansi_0.5.0      
[49] lifecycle_1.0.1   stringi_1.7.6     whisker_0.4       yaml_2.2.1       
[53] grid_4.1.1        parallel_4.1.1    listenv_0.8.0     promises_1.2.0.1 
[57] crayon_1.4.2      lattice_0.20-45   haven_2.4.3       hms_1.1.1        
[61] tmvnsim_1.0-2     knitr_1.37        ps_1.6.0          pillar_1.6.4     
[65] codetools_0.2-18  reprex_2.0.1      glue_1.6.0        evaluate_0.14    
[69] getPass_0.2-2     modelr_0.1.8      vctrs_0.3.8       tzdb_0.2.0       
[73] httpuv_1.6.5      cellranger_1.1.0  gtable_0.3.0      assertthat_0.2.1 
[77] xfun_0.29         broom_0.7.11      later_1.3.0       globals_0.14.0   
[81] ellipsis_0.3.2    here_1.0.1       </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
