<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="wolfemd" />

<meta name="date" content="2022-05-09" />

<title>Predict crosses</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GSexample2022</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="download.html">Download</a>
</li>
<li>
  <a href="clean_phenos.html">Clean phenos</a>
</li>
<li>
  <a href="prep_genos.html">Prep genos</a>
</li>
<li>
  <a href="get_blups.html">Get BLUPs</a>
</li>
<li>
  <a href="intro_gp.html">Intro to GP</a>
</li>
<li>
  <a href="kfoldcrossval.html">K-fold Cross-val</a>
</li>
<li>
  <a href="get_gebv.html">Get GEBVs</a>
</li>
<li>
  <a href="parentwise_cv.html">Parent-wise CV</a>
</li>
<li>
  <a href="pred_crosses.html">Predict crosses</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Predict crosses</h1>
<h4 class="author">wolfemd</h4>
<h4 class="date">2022-05-09</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-05-09
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>GSexample2022/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220320code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20220320)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220320code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220320)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomwolfemdGSexample2022treebb2d98179db87acbe2189a400bac5d598f9119d8targetblankbb2d981a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/wolfemd/GSexample2022/tree/bb2d98179db87acbe2189a400bac5d598f9119d8" target="_blank">bb2d981</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomwolfemdGSexample2022treebb2d98179db87acbe2189a400bac5d598f9119d8targetblankbb2d981a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/wolfemd/GSexample2022/tree/bb2d98179db87acbe2189a400bac5d598f9119d8" target="_blank">bb2d981</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  .DS_Store
    Untracked:  Icon
    Untracked:  analysis/.DS_Store
    Untracked:  analysis/BreedBaseGenotypes_subset.log
    Untracked:  analysis/Icon
    Untracked:  analysis/images/Icon
    Untracked:  analysis/images/paste-2F34C15B.png
    Untracked:  analysis/images/paste-AD7AEEAC.png
    Untracked:  analysis/images/paste-CB5D91AC.png
    Untracked:  analysis/troubleshooting.Rmd
    Untracked:  code/Icon
    Untracked:  data/.DS_Store
    Untracked:  data/BreedBaseGenotypesDownload.positions
    Untracked:  data/BreedBaseGenotypesDownload.vcf
    Untracked:  data/BreedBaseGenotypes_subset.bed
    Untracked:  data/BreedBaseGenotypes_subset.bim
    Untracked:  data/BreedBaseGenotypes_subset.fam
    Untracked:  data/BreedBaseGenotypes_subset.hap.gz
    Untracked:  data/BreedBaseGenotypes_subset.log
    Untracked:  data/BreedBaseGenotypes_subset.nosex
    Untracked:  data/BreedBaseGenotypes_subset.positions
    Untracked:  data/BreedBaseGenotypes_subset.samples
    Untracked:  data/Icon
    Untracked:  data/metadata_cynthia.csv
    Untracked:  data/phenotype_cynthia.csv
    Untracked:  data/subset_unique_names_for_vcf.txt
    Untracked:  data/unique_names_for_vcf.txt
    Untracked:  data/vcf_colnames.txt
    Untracked:  out.log
    Untracked:  output/.DS_Store
    Untracked:  output/BreedBaseGenotypes_subset.genome
    Untracked:  output/BreedBaseGenotypes_subset.log
    Untracked:  output/BreedBaseGenotypes_subset.nosex
    Untracked:  output/Icon
    Untracked:  output/genomicPredictions.rds
    Untracked:  output/models_compared.Rdata
    Untracked:  output/parentWiseCV.rds
    Untracked:  output/standardCV.rds
    Untracked:  output/verified_ped.rds

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/pred_crosses.Rmd</code>) and HTML (<code>docs/pred_crosses.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/wolfemd/GSexample2022/blob/bb2d98179db87acbe2189a400bac5d598f9119d8/analysis/pred_crosses.Rmd" target="_blank">bb2d981</a>
</td>
<td>
wolfemd
</td>
<td>
2022-05-09
</td>
<td>
Add the final two sections - parent-wise cross-validation plus predicting crosses
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="load-packages" class="section level1">
<h1>Load packages</h1>
<pre class="r"><code>suppressMessages(library(tidyverse)); 
suppressMessages(library(genomicMateSelectR)); </code></pre>
</div>
<div id="set-up" class="section level1">
<h1>Set-up</h1>
<pre class="r"><code># GENOMIC RELATIONSHIP MATRIX
grms&lt;-list(A=readRDS(file=here::here(&quot;output&quot;,&quot;kinship_add.rds&quot;)))

# BLUPs
blups&lt;-readRDS(here::here(&quot;output&quot;,&quot;blups.rds&quot;)) %&gt;% 
     # based on cross-validation, decided to exclude MCMDS from this analysis
     filter(Trait != &quot;MCMDS&quot;) %&gt;% 
     # need to rename the &quot;blups&quot; list to comply with the runCrossVal function
     rename(TrainingData=blups) %&gt;% 
     dplyr::select(Trait,TrainingData) %&gt;% 
     # need also to remove phenotyped-but-not-genotyped lines
     mutate(TrainingData=map(TrainingData,
                             ~filter(.,germplasmName %in% rownames(grms$A)) %&gt;% 
                                  # rename the germplasmName column to GID
                                  rename(GID=germplasmName)))

# DOSAGE MATRIX
## Dosages are also needed for runGenomicPredictions() when getMarkEffs=TRUE
## Reason is that they are used to extra SNP effects from GBLUP models
dosages&lt;-readRDS(here::here(&quot;data&quot;,&quot;dosages.rds&quot;))

# SELECTION INDEX
SIwts&lt;-c(DM=15,
         logFYLD=20,
         logDYLD=20)</code></pre>
</div>
<div id="get-marker-effects" class="section level1">
<h1>Get marker effects</h1>
<p>We need to re-run <code>runGenomicPredictions()</code>, this time using the <code>getMarkEffs=TRUE</code> option.</p>
<p>Will “backsolve” the RR-BLUP marker effect solutions from the GBLUP solutions, using the <code>backsolveSNPeff()</code> function under-the-hood.</p>
<pre class="r"><code>gpreds_withMarkEffs&lt;-runGenomicPredictions(modelType = &quot;A&quot;, 
                                           selInd = T, SIwts = SIwts,
                                           getMarkEffs = TRUE,
                                           dosages = dosages,
                                           blups = blups, 
                                           grms = grms,
                                           ncores=3)</code></pre>
<pre><code>Loading required package: furrr</code></pre>
<pre><code>Loading required package: future</code></pre>
<pre><code>iteration    LogLik     wall    cpu(sec)   restrained
    1      -356.471   20:26:31      4           0
    2      -355.724   20:26:34      7           0
    3      -355.405   20:26:37      10           0
    4      -355.333   20:26:40      13           0
    5      -355.327   20:26:42      15           0
    6      -355.326   20:26:45      18           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -135.425   20:26:28      0           0
    2      -135.407   20:26:28      0           0
    3      -135.395   20:26:28      0           0
    4      -135.391   20:26:28      0           0
    5      -135.39   20:26:28      0           0
iteration    LogLik     wall    cpu(sec)   restrained
    1      -146.037   20:26:28      0           0
    2      -146.031   20:26:29      1           0
    3      -146.028   20:26:29      1           0
    4      -146.027   20:26:29      1           0</code></pre>
<pre class="r"><code>saveRDS(gpreds_withMarkEffs,
        file = here::here(&quot;output&quot;,&quot;genomicPredictions_withMarkEffs.rds&quot;))</code></pre>
<p>Notice that there is now a additional list-type column with the label “allelesubsnpeff” indicating that, because we ran an additive-only model, these SNP-effects represent predictions of allele substitution effects.</p>
<pre class="r"><code>gpreds_withMarkEffs$genomicPredOut[[1]]</code></pre>
<pre><code># A tibble: 3 × 5
  Trait   gblups               varcomps     fixeffs      allelesubsnpeff  
  &lt;chr&gt;   &lt;list&gt;               &lt;list&gt;       &lt;list&gt;       &lt;list&gt;           
1 DM      &lt;tibble [1,007 × 2]&gt; &lt;df [2 × 4]&gt; &lt;df [1 × 5]&gt; &lt;dbl [3,987 × 1]&gt;
2 logFYLD &lt;tibble [1,007 × 2]&gt; &lt;df [2 × 4]&gt; &lt;df [1 × 5]&gt; &lt;dbl [3,987 × 1]&gt;
3 logDYLD &lt;tibble [1,007 × 2]&gt; &lt;df [2 × 4]&gt; &lt;df [1 × 5]&gt; &lt;dbl [3,987 × 1]&gt;</code></pre>
<pre class="r"><code>gpreds_withMarkEffs$genomicPredOut[[1]]$allelesubsnpeff[[1]][1:5,]</code></pre>
<pre><code> 1_652699_G_C  1_868970_G_T  1_943129_T_A 1_1132830_A_T 1_1310706_A_T 
-0.0035035330  0.0001138839  0.0002113849  0.0027187074 -0.0041179418 </code></pre>
</div>
<div id="crosses-to-predict" class="section level1">
<h1>Crosses-to-predict</h1>
<p>Will also need a data.frame indicating what pairs of parents we want to predict crosses for.</p>
<p>As a convenience, we can use the <code>crosses2predict()</code> function. Makes a data.frame from a vector of genotype ID’s.</p>
<p>A realistic approach, is to choose a set of parents based on their GEBV, but more than we’d actually like to actually make crosses with.</p>
<p>It is still somewhat computationally intensive to predict the variances and covariances of traits for each cross, so we can’t quite predict all possible pairwise crosses… definitely not on our laptops for this example.</p>
<p>Here, as an example, picking the top 10 candidate parents:</p>
<pre class="r"><code># Access the predicted GEBV
top10parents&lt;-gpreds_withMarkEffs$gblups[[1]] %&gt;% 
     # Arrange in descending order based on the SELIND
     arrange(desc(SELIND)) %&gt;% 
     # I&#39;ll pick the top 10 parents
     slice(1:10) %$%
     # And extract their GID to a vector
     GID</code></pre>
<pre class="r"><code>CrossesToPredict&lt;-crosses2predict(top10parents)</code></pre>
<pre class="r"><code>CrossesToPredict %&gt;% head</code></pre>
<pre><code># A tibble: 6 × 2
  sireID      damID             
  &lt;chr&gt;       &lt;chr&gt;             
1 TMEB117(4X) TMEB117(4X)       
2 TMEB117(4X) TMEB693           
3 TMEB117(4X) TMEB789           
4 TMEB117(4X) IITA-TMS-IBA970103
5 TMEB117(4X) IITA-TMS-IBA970255
6 TMEB117(4X) NANCHINYAYA       </code></pre>
<pre class="r"><code>CrossesToPredict %&gt;% nrow()</code></pre>
<pre><code>[1] 55</code></pre>
</div>
<div id="run-predictcrosses" class="section level1">
<h1>Run <code>predictCrosses()</code></h1>
<p>Additional inputs we will need: “haplotype matrix” and “recombination frequency matrix.”</p>
<pre class="r"><code># HAPLOTYPE MATRIX
## keep only haplos for candidate parents we want to predict crosses for
## those which will be used in prediction, saves memory
haploMat&lt;-readRDS(file=here::here(&quot;data&quot;,&quot;haplotypes.rds&quot;))
parenthaps&lt;-sort(c(paste0(top10parents,&quot;_HapA&quot;),
                   paste0(top10parents,&quot;_HapB&quot;)))
haploMat&lt;-haploMat[parenthaps,]

# RECOMBINATION FREQUENCY MATRIX
recombFreqMat&lt;-readRDS(file=here::here(&quot;output&quot;,&quot;recombFreqMat_1minus2c.rds&quot;))</code></pre>
<p>Let’s go!</p>
<pre class="r"><code>starttime&lt;-proc.time()[3]
crossPreds&lt;-predictCrosses(modelType=&quot;A&quot;,
                           selInd = T, SIwts = SIwts,
                           CrossesToPredict=CrossesToPredict,
                           snpeffs=gpreds_withMarkEffs$genomicPredOut[[1]], 
                           haploMat=haploMat,
                           dosages = dosages[top10parents,],
                           recombFreqMat=recombFreqMat,
                           ncores=10)</code></pre>
<pre><code>[1] &quot;Predicting cross variance parameters&quot;
[1] &quot;Done predicting fam vars. Took 0.22 mins for 54 crosses&quot;
[1] &quot;Predicting cross means&quot;
[1] &quot;Computing SELECTION INDEX means and variances.&quot;</code></pre>
<pre class="r"><code>elapsed&lt;-proc.time()[3]-starttime; elapsed/60</code></pre>
<pre><code>elapsed 
 0.3935 </code></pre>
<p>Save results</p>
<pre class="r"><code>saveRDS(crossPreds,file = here::here(&quot;output&quot;,&quot;predictedCrosses.rds&quot;))</code></pre>
</div>
<div id="select-crosses" class="section level1">
<h1>Select crosses</h1>
<pre class="r"><code>crossPreds&lt;-readRDS(here::here(&quot;output&quot;,&quot;predictedCrosses.rds&quot;))
crossPreds</code></pre>
<pre><code># A tibble: 1 × 2
  tidyPreds          rawPreds        
  &lt;list&gt;             &lt;list&gt;          
1 &lt;tibble [216 × 9]&gt; &lt;named list [2]&gt;</code></pre>
<pre class="r"><code>crossPreds$tidyPreds[[1]] %&gt;% str</code></pre>
<pre><code>tibble [216 × 9] (S3: tbl_df/tbl/data.frame)
 $ sireID        : chr [1:216] &quot;TMEB117(4X)&quot; &quot;TMEB117(4X)&quot; &quot;TMEB117(4X)&quot; &quot;TMEB117(4X)&quot; ...
 $ damID         : chr [1:216] &quot;TMEB693&quot; &quot;TMEB693&quot; &quot;TMEB693&quot; &quot;TMEB693&quot; ...
 $ Nsegsnps      : int [1:216] 1663 1663 1663 1663 1669 1669 1669 1669 1756 1756 ...
 $ predOf        : chr [1:216] &quot;BV&quot; &quot;BV&quot; &quot;BV&quot; &quot;BV&quot; ...
 $ Trait         : chr [1:216] &quot;SELIND&quot; &quot;DM&quot; &quot;logFYLD&quot; &quot;logDYLD&quot; ...
 $ predMean      : num [1:216] 17.6893 1.1104 0.0116 0.04 17.3643 ...
 $ predVar       : num [1:216] 1.96e+01 8.62e-02 1.56e-04 1.49e-04 1.15e+01 ...
 $ predSD        : num [1:216] 4.4264 0.2937 0.0125 0.0122 3.3915 ...
 $ predUsefulness: num [1:216] 29.5078 1.8945 0.0449 0.0727 26.4196 ...</code></pre>
<p>Output of predictCrosses() is a tibble. Two columns, 1 row. Column 1 (tidyPreds) has cleaned-up “tidy” predictions. Column 2 (rawPreds) has more detailed output.</p>
<p>Remember that the “usefulness” (predUsefulness) here, or UC for short, is equal to a prediction of the expected mean of the top fraction of progeny from each cross</p>
<p><span class="math display">\[\hat{UC} = \hat{\mu} + i \times \hat{\sigma}\]</span></p>
<p>This is also called the “superior progeny mean” in the literature.</p>
<p>Actually, the user has the option to modify the expected, standardized selection intensity</p>
<p>either in advance with the <code>stdSelInt=</code> argument to <code>predictCrosses()</code> or after-the-fact;</p>
<p>The default is a value of 2.67, corresponding to selecting the top 1% of offspring for the cross you will be making.</p>
<p>So let’s say you want to make only the top 10 of the 55 predicted crosses:</p>
<pre class="r"><code>top10crosses&lt;-crossPreds$tidyPreds[[1]] %&gt;% 
     filter(Trait==&quot;SELIND&quot;) %&gt;% 
     dplyr::select(-predVar) %&gt;% 
     arrange(desc(predUsefulness)) %&gt;% 
     slice(1:10)
top10crosses</code></pre>
<pre><code># A tibble: 10 × 8
   sireID          damID    Nsegsnps predOf Trait predMean predSD predUsefulness
   &lt;chr&gt;           &lt;chr&gt;       &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;dbl&gt;  &lt;dbl&gt;          &lt;dbl&gt;
 1 TMEB693         TMEB693      1041 BV     SELI…     17.7   4.45           29.5
 2 TMEB117(4X)     TMEB693      1663 BV     SELI…     17.7   4.43           29.5
 3 TMEB693         TMS14F1…     2327 BV     SELI…     15.5   4.87           28.5
 4 TMEB693         TMEB789      1682 BV     SELI…     17.3   4.19           28.5
 5 TMEB693         IITA-TM…     2083 BV     SELI…     16.9   4.27           28.3
 6 TMEB693         IITA-TM…     2077 BV     SELI…     17.0   4.25           28.3
 7 TMEB693         NANCHIN…     1699 BV     SELI…     16.3   4.25           27.6
 8 TMEB789         TMS14F1…     2398 BV     SELI…     15.2   4.63           27.6
 9 TMEB789         TMEB789      1157 BV     SELI…     17.0   3.91           27.5
10 TMS14F1035P0004 TMS14F1…     1429 BV     SELI…     13.4   5.26           27.4</code></pre>
<p>And thus we have a crossing plan! Congratulations!</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.1 (2021-08-10)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] furrr_0.2.3              future_1.23.0            genomicMateSelectR_0.2.0
 [4] forcats_0.5.1            stringr_1.4.0            dplyr_1.0.7             
 [7] purrr_0.3.4              readr_2.1.1              tidyr_1.1.4             
[10] tibble_3.1.6             ggplot2_3.3.5            tidyverse_1.3.1         
[13] workflowr_1.7.0         

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.7        here_1.0.1        lubridate_1.8.0   listenv_0.8.0    
 [5] getPass_0.2-2     ps_1.6.0          assertthat_0.2.1  rprojroot_2.0.2  
 [9] digest_0.6.29     utf8_1.2.2        parallelly_1.30.0 R6_2.5.1         
[13] cellranger_1.1.0  backports_1.4.1   reprex_2.0.1      evaluate_0.14    
[17] httr_1.4.2        pillar_1.6.4      rlang_0.4.12      readxl_1.3.1     
[21] rstudioapi_0.13   whisker_0.4       callr_3.7.0       jquerylib_0.1.4  
[25] rmarkdown_2.11    munsell_0.5.0     broom_0.7.11      compiler_4.1.1   
[29] httpuv_1.6.5      modelr_0.1.8      xfun_0.29         pkgconfig_2.0.3  
[33] globals_0.14.0    htmltools_0.5.2   tidyselect_1.1.1  codetools_0.2-18 
[37] fansi_0.5.0       crayon_1.4.2      tzdb_0.2.0        dbplyr_2.1.1     
[41] withr_2.4.3       later_1.3.0       grid_4.1.1        jsonlite_1.7.2   
[45] gtable_0.3.0      lifecycle_1.0.1   DBI_1.1.2         git2r_0.29.0     
[49] magrittr_2.0.1    scales_1.1.1      cli_3.1.0         stringi_1.7.6    
[53] fs_1.5.2          promises_1.2.0.1  xml2_1.3.3        bslib_0.3.1      
[57] ellipsis_0.3.2    generics_0.1.1    vctrs_0.3.8       tools_4.1.1      
[61] glue_1.6.0        hms_1.1.1         parallel_4.1.1    processx_3.5.2   
[65] fastmap_1.1.0     yaml_2.2.1        colorspace_2.0-2  rvest_1.0.2      
[69] knitr_1.37        haven_2.4.3       sass_0.4.0       </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
